// @polygonjs/polygonjs/dist/src/engine/poly/registers/nodes/types/Cop.js
"use strict";
export var CopType = /* @__PURE__ */ ((CopType2) => {
  CopType2["BUILDER"] = "builder";
  CopType2["BUILDER_2D_ARRAY"] = "builder2DArray";
  CopType2["CUBE_MAP"] = "cubeMap";
  CopType2["CUBE_CAMERA"] = "cubeCamera";
  CopType2["CUBE_MAP_FROM_SCENE"] = "cubeMapFromScene";
  CopType2["ENV_MAP"] = "envMap";
  CopType2["GIF"] = "gif";
  CopType2["LIGHT_MAP"] = "lightMap";
  CopType2["LUT"] = "lut";
  CopType2["RENDER"] = "render";
  CopType2["SDF_BLUR"] = "SDFBlur";
  CopType2["SDF_FROM_OBJECT"] = "SDFFromObject";
  CopType2["SDF_FROM_URL"] = "SDFFromUrl";
  CopType2["SNAPSHOT"] = "snapshot";
  CopType2["TEXT"] = "text";
  CopType2["VIDEO"] = "video";
  CopType2["WEB_CAM"] = "webCam";
  return CopType2;
})(CopType || {});
export var CopTypeImage = /* @__PURE__ */ ((CopTypeImage2) => {
  CopTypeImage2["IMAGE"] = "image";
  CopTypeImage2["IMAGE_EXR"] = "imageEXR";
  CopTypeImage2["IMAGE_HDR"] = "imageHDR";
  CopTypeImage2["IMAGE_KTX2"] = "imageKTX2";
  return CopTypeImage2;
})(CopTypeImage || {});
