// @polygonjs/polygonjs/dist/src/engine/poly/registers/nodes/types/Gl.js
"use strict";
export var GlType = /* @__PURE__ */ ((GlType2) => {
  GlType2["ADJACENT_POINTS_ATTRIB_SMOOTH"] = "adjacentPointsAttribSmooth";
  GlType2["ADJACENT_UV_ATTRIB_SMOOTH"] = "adjacentUvAttribSmooth";
  GlType2["ATTRIBUTE"] = "attribute";
  GlType2["COMPUTE_NORMALS"] = "computeNormals";
  GlType2["CONSTANT"] = "constant";
  GlType2["FOR_LOOP"] = "forLoop";
  GlType2["IF_THEN"] = "ifThen";
  GlType2["GLOBALS"] = "globals";
  GlType2["NOISE"] = "noise";
  GlType2["OUTPUT"] = "output";
  GlType2["PARAM"] = "param";
  GlType2["RAMP"] = "ramp";
  GlType2["SDF_2D_BOX"] = "SDF2DBox";
  GlType2["SDF_2D_CIRCLE"] = "SDF2DCircle";
  GlType2["SDF_2D_CROSS"] = "SDF2DCross";
  GlType2["SDF_2D_HEART"] = "SDF2DHeart";
  GlType2["SDF_2D_ROUNDED_X"] = "SDF2DRoundedX";
  GlType2["SDF_2D_STAIRS"] = "SDF2DStairs";
  GlType2["SDF_BOX"] = "SDFBox";
  GlType2["SDF_BOX_FRAME"] = "SDFBoxFrame";
  GlType2["SDF_BOX_ROUND"] = "SDFBoxRound";
  GlType2["SDF_CAPSULE"] = "SDFCapsule";
  GlType2["SDF_CAPSULE_VERTICAL"] = "SDFCapsuleVertical";
  GlType2["SDF_CONE"] = "SDFCone";
  GlType2["SDF_CONE_ROUND"] = "SDFConeRound";
  GlType2["SDF_ELONGATE"] = "SDFElongate";
  GlType2["SDF_FRACTAL_MANDELBROT"] = "SDFFractalMandelbrot";
  GlType2["SDF_GRADIENT"] = "SDFGradient";
  GlType2["SDF_HEXAGONAL_PRISM"] = "SDFHexagonalPrism";
  GlType2["SDF_HORSE_SHOE"] = "SDFHorseShoe";
  GlType2["SDF_LINK"] = "SDFLink";
  GlType2["SDF_OCTAHEDRON"] = "SDFOctahedron";
  GlType2["SDF_OCTOGONAL_PRISM"] = "SDFOctogonalPrism";
  GlType2["SDF_ONION"] = "SDFOnion";
  GlType2["SDF_PLANE"] = "SDFPlane";
  GlType2["SDF_PYRAMID"] = "SDFPyramid";
  GlType2["SDF_QUAD"] = "SDFQuad";
  GlType2["SDF_REPEAT_POLAR"] = "SDFRepeatPolar";
  GlType2["SDF_REVOLUTION"] = "SDFRevolution";
  GlType2["SDF_RHOMBUS"] = "SDFRhombus";
  GlType2["SDF_RHOMBUS_TRIACONTAHEDRON"] = "SDFRhombusTriacontahedron";
  GlType2["SDF_SOLID_ANGLE"] = "SDFSolidAngle";
  GlType2["SDF_SPHERE"] = "SDFSphere";
  GlType2["SDF_SPHERE_CUT"] = "SDFSphereCut";
  GlType2["SDF_SPHERE_HOLLOW"] = "SDFSphereHollow";
  GlType2["SDF_TORUS"] = "SDFTorus";
  GlType2["SDF_TRANSFORM"] = "SDFTransform";
  GlType2["SDF_TRIANGULAR_PRISM"] = "SDFTriangularPrism";
  GlType2["SDF_TRIANGLE"] = "SDFTriangle";
  GlType2["SDF_TUBE"] = "SDFTube";
  GlType2["SDF_TWIST"] = "SDFTwist";
  GlType2["SDF_CONTEXT"] = "SDFContext";
  GlType2["SDF_MATERIAL"] = "SDFMaterial";
  GlType2["TEXTURE"] = "texture";
  GlType2["TEXTURE_DISPLACEMENT"] = "textureDisplacement";
  GlType2["TEXTURE_SDF"] = "textureSDF";
  GlType2["VARYING_READ"] = "varyingRead";
  GlType2["VARYING_WRITE"] = "varyingWrite";
  GlType2["VERTEX_ANIMATION_TEXTURE"] = "vertexAnimationTexture";
  GlType2["VERTEX_ANIMATION_TEXTURE_INTERPOLATED"] = "vertexAnimationTextureInterpolated";
  return GlType2;
})(GlType || {});
