// @polygonjs/polygonjs/dist/src/engine/poly/registers/nodes/types/Js.js
"use strict";
export var JsType = /* @__PURE__ */ ((JsType2) => {
  JsType2["ANIMATION_MIXER"] = "animationMixer";
  JsType2["ANIMATION_MIXER_UPDATE"] = "animationMixerUpdate";
  JsType2["ATTRIBUTE"] = "attribute";
  JsType2["CODE"] = "code";
  JsType2["CURSOR"] = "cursor";
  JsType2["FOR_LOOP"] = "forLoop";
  JsType2["GEOLOCATION_CURRENT_POSITION"] = "geolocationCurrentPosition";
  JsType2["GET_INSTANCE_ATTRIBUTE"] = "getInstanceAttribute";
  JsType2["GET_INSTANCE_PROPERTY"] = "getInstanceProperty";
  JsType2["GET_OBJECT_ATTRIBUTE"] = "getObjectAttribute";
  JsType2["GET_OBJECT_PROPERTY"] = "getObjectProperty";
  JsType2["GET_OBJECT_USER_DATA"] = "getObjectUserData";
  JsType2["GET_NODE"] = "getNode";
  JsType2["GET_PARAM"] = "getParam";
  JsType2["GET_POINT_ATTRIBUTE"] = "getPointAttribute";
  JsType2["GET_POINT_PROPERTY"] = "getPointProperty";
  JsType2["GET_PRIMITIVE_ATTRIBUTE"] = "getPrimitiveAttribute";
  JsType2["GLOBALS"] = "globals";
  JsType2["GLOBALS_AMBIENT_LIGHT"] = "globalsAmbientLight";
  JsType2["GLOBALS_AREA_LIGHT"] = "globalsAreaLight";
  JsType2["GLOBALS_DIRECTIONAL_LIGHT"] = "globalsDirectionalLight";
  JsType2["GLOBALS_HEMISPHERE_LIGHT"] = "globalsHemisphereLight";
  JsType2["GLOBALS_POINT_LIGHT"] = "globalsPointLight";
  JsType2["GLOBALS_SPOT_LIGHT"] = "globalsSpotLight";
  JsType2["IMPORT_ATTRIBUTE"] = "importAttribute";
  JsType2["KEYFRAMES"] = "keyframes";
  JsType2["ON_CHILD_ATTRIBUTE_UPDATE"] = "onChildAttributeUpdate";
  JsType2["ON_KEY"] = "onKey";
  JsType2["ON_KEYDOWN"] = "onKeydown";
  JsType2["ON_KEYPRESS"] = "onKeypress";
  JsType2["ON_KEYUP"] = "onKeyup";
  JsType2["ON_MAPBOX_CAMERA_MOVE"] = "onMapboxCameraMove";
  JsType2["ON_MAPBOX_CAMERA_MOVE_START"] = "onMapboxCameraMoveStart";
  JsType2["ON_MAPBOX_CAMERA_MOVE_END"] = "onMapboxCameraMoveEnd";
  JsType2["ON_OBJECT_ATTRIBUTE_UPDATE"] = "onObjectAttributeUpdate";
  JsType2["ON_OBJECT_BEFORE_DELETE"] = "onObjectBeforeDelete";
  JsType2["ON_OBJECT_CLICK"] = "onObjectClick";
  JsType2["ON_OBJECT_CLICK_GPU"] = "onObjectClickGPU";
  JsType2["ON_OBJECT_CONTEXT_MENU"] = "onObjectContextMenu";
  JsType2["ON_OBJECT_CONTEXT_MENU_GPU"] = "onObjectContextMenuGPU";
  JsType2["ON_OBJECT_DISPATCH_EVENT"] = "onObjectDispatchEvent";
  JsType2["ON_OBJECT_HOVER"] = "onObjectHover";
  JsType2["ON_OBJECT_HOVER_GPU"] = "onObjectHoverGPU";
  JsType2["ON_OBJECT_LONG_PRESS"] = "onObjectLongPress";
  JsType2["ON_OBJECT_LONG_PRESS_GPU"] = "onObjectLongPressGPU";
  JsType2["ON_OBJECT_MOUSE_CLICK"] = "onObjectMouseClick";
  JsType2["ON_OBJECT_POINTERDOWN"] = "onObjectPointerdown";
  JsType2["ON_OBJECT_POINTERDOWN_GPU"] = "onObjectPointerdownGPU";
  JsType2["ON_OBJECT_POINTERUP"] = "onObjectPointerup";
  JsType2["ON_OBJECT_POINTERUP_GPU"] = "onObjectPointerupGPU";
  JsType2["ON_OBJECT_SWIPE"] = "onObjectSwipe";
  JsType2["ON_OBJECT_SWIPE_GPU"] = "onObjectSwipeGPU";
  JsType2["ON_OBJECT_SWIPEDOWN"] = "onObjectSwipedown";
  JsType2["ON_OBJECT_SWIPELEFT"] = "onObjectSwipeleft";
  JsType2["ON_OBJECT_SWIPERIGHT"] = "onObjectSwiperight";
  JsType2["ON_OBJECT_SWIPEUP"] = "onObjectSwipeup";
  JsType2["ON_POINTERDOWN"] = "onPointerdown";
  JsType2["ON_POINTERUP"] = "onPointerup";
  JsType2["ON_SCENE_PAUSE"] = "onScenePause";
  JsType2["ON_SCENE_PLAY"] = "onScenePlay";
  JsType2["ON_SCENE_RESET"] = "onSceneReset";
  JsType2["ON_SCROLL"] = "onScroll";
  JsType2["ON_MANUAL_TRIGGER"] = "onManualTrigger";
  JsType2["ON_TICK"] = "onTick";
  JsType2["ON_PERFORMANCE_CHANGE"] = "onPerformanceChange";
  JsType2["ON_VIDEO_EVENT"] = "onVideoEvent";
  JsType2["ON_VIEWER_CONTROLS_EVENT"] = "onViewerControlsEvent";
  JsType2["ON_WEBXR_CONTROLLER_EVENT"] = "onWebXRControllerEvent";
  JsType2["OUTPUT"] = "output";
  JsType2["OUTPUT_AMBIENT_LIGHT"] = "outputAmbientLight";
  JsType2["OUTPUT_AREA_LIGHT"] = "outputAreaLight";
  JsType2["OUTPUT_DIRECTIONAL_LIGHT"] = "outputDirectionalLight";
  JsType2["OUTPUT_HEMISPHERE_LIGHT"] = "outputHemisphereLight";
  JsType2["OUTPUT_POINT_LIGHT"] = "outputPointLight";
  JsType2["OUTPUT_SPOT_LIGHT"] = "outputSpotLight";
  JsType2["PARAM"] = "param";
  JsType2["RAY_FROM_CURSOR"] = "rayFromCursor";
  JsType2["RAY_FROM_CAMERA"] = "rayFromCamera";
  JsType2["SDF_2D_BOX"] = "SDF2DBox";
  JsType2["SDF_2D_CIRCLE"] = "SDF2DCircle";
  JsType2["SDF_2D_CROSS"] = "SDF2DCross";
  JsType2["SDF_2D_HEART"] = "SDF2DHeart";
  JsType2["SDF_2D_ROUNDED_X"] = "SDF2DRoundedX";
  JsType2["SDF_BOX"] = "SDFBox";
  JsType2["SDF_ELONGATE"] = "SDFElongate";
  JsType2["SDF_ONION"] = "SDFOnion";
  JsType2["SDF_PLANE"] = "SDFPlane";
  JsType2["SDF_REVOLUTION"] = "SDFRevolution";
  JsType2["SDF_SPHERE"] = "SDFSphere";
  JsType2["SDF_TORUS"] = "SDFTorus";
  JsType2["SDF_TRANSFORM"] = "SDFTransform";
  JsType2["SDF_TUBE"] = "SDFTube";
  JsType2["SET_INSTANCE_ATTRIBUTE"] = "setInstanceAttribute";
  JsType2["SET_INSTANCE_LOOK_AT"] = "setInstanceLookAt";
  JsType2["SET_INSTANCE_POSITION"] = "setInstancePosition";
  JsType2["SET_INSTANCE_QUATERNION"] = "setInstanceQuaternion";
  JsType2["SET_INSTANCE_SCALE"] = "setInstanceScale";
  JsType2["SET_OBJECT_ATTRIBUTE"] = "setObjectAttribute";
  JsType2["SET_OBJECT_LOOK_AT"] = "setObjectLookAt";
  JsType2["SET_OBJECT_POSITION"] = "setObjectPosition";
  JsType2["SET_OBJECT_QUATERNION"] = "setObjectQuaternion";
  JsType2["SET_OBJECT_ROTATION"] = "setObjectRotation";
  JsType2["SET_OBJECT_SCALE"] = "setObjectScale";
  JsType2["SET_PARAM"] = "setParam";
  JsType2["SET_PHYSICS_RBD_POSITION"] = "setPhysicsRBDPosition";
  JsType2["SET_PLAYER_INPUT"] = "setPlayerInput";
  JsType2["SET_POINT_ATTRIBUTE"] = "setPointAttribute";
  JsType2["SET_POINT_POSITION"] = "setPointPosition";
  JsType2["SET_SOFT_BODY_CONSTRAINT_POSITION"] = "setSoftBodyConstraintPosition";
  JsType2["SET_SOFT_BODY_POSITION"] = "setSoftBodyPosition";
  JsType2["SET_SOFT_BODY_VELOCITY"] = "setSoftBodyVelocity";
  JsType2["SET_WFC_SOFT_CONSTRAINT"] = "setWFCSoftConstraint";
  JsType2["WFC_BUILD"] = "WFCBuild";
  return JsType2;
})(JsType || {});
