// @polygonjs/polygonjs/dist/src/engine/nodes/js/code/assemblers/pointBuilder/PointBuilderAssemblerCommon.js
"use strict";
export var PointBuilderAssemblerConstant = /* @__PURE__ */ ((PointBuilderAssemblerConstant2) => {
  PointBuilderAssemblerConstant2["POINT_CONTAINER"] = "pointContainer";
  PointBuilderAssemblerConstant2["POSITION"] = "pointContainer.position";
  PointBuilderAssemblerConstant2["NORMAL"] = "pointContainer.normal";
  PointBuilderAssemblerConstant2["PTNUM"] = "pointContainer.ptnum";
  PointBuilderAssemblerConstant2["OBJNUM"] = "pointContainer.objnum";
  PointBuilderAssemblerConstant2["NORMALS_UPDATED"] = "pointContainer.normalsUpdated";
  PointBuilderAssemblerConstant2["ATTRIBUTES_DICT"] = "attributesDict";
  PointBuilderAssemblerConstant2["OBJECT_3D"] = "null";
  PointBuilderAssemblerConstant2["MATERIAL"] = "null";
  PointBuilderAssemblerConstant2["PRIMITIVE_GRAPH"] = "null";
  return PointBuilderAssemblerConstant2;
})(PointBuilderAssemblerConstant || {});
export var PointVariable = /* @__PURE__ */ ((PointVariable2) => {
  PointVariable2["POSITION"] = "position";
  PointVariable2["NORMAL"] = "normal";
  PointVariable2["PTNUM"] = "ptnum";
  PointVariable2["OBJNUM"] = "objnum";
  return PointVariable2;
})(PointVariable || {});
