// @polygonjs/polygonjs/dist/src/engine/expressions/methods/Common.js
"use strict";
export var ExpressionType = /* @__PURE__ */ ((ExpressionType2) => {
  ExpressionType2["animationNames"] = "animationNames";
  ExpressionType2["arg"] = "arg";
  ExpressionType2["argc"] = "argc";
  ExpressionType2["bbox"] = "bbox";
  ExpressionType2["blob"] = "blob";
  ExpressionType2["cameraName"] = "cameraName";
  ExpressionType2["cameraNames"] = "cameraNames";
  ExpressionType2["camerasCount"] = "camerasCount";
  ExpressionType2["centroid"] = "centroid";
  ExpressionType2["ch"] = "ch";
  ExpressionType2["chsop"] = "chsop";
  ExpressionType2["copy"] = "copy";
  ExpressionType2["copRes"] = "copRes";
  ExpressionType2["isDeviceMobile"] = "isDeviceMobile";
  ExpressionType2["isDeviceTouch"] = "isDeviceTouch";
  ExpressionType2["len"] = "len";
  ExpressionType2["join"] = "join";
  ExpressionType2["js"] = "js";
  ExpressionType2["object"] = "object";
  ExpressionType2["objectsCount"] = "objectsCount";
  ExpressionType2["objectName"] = "objectName";
  ExpressionType2["objectNames"] = "objectNames";
  ExpressionType2["opdigits"] = "opdigits";
  ExpressionType2["opname"] = "opname";
  ExpressionType2["padzero"] = "padzero";
  ExpressionType2["playerMode"] = "playerMode";
  ExpressionType2["point"] = "point";
  ExpressionType2["pointsCount"] = "pointsCount";
  ExpressionType2["smoothstep"] = "smoothstep";
  ExpressionType2["smootherstep"] = "smootherstep";
  ExpressionType2["solverIteration"] = "solverIteration";
  ExpressionType2["strCharsCount"] = "strCharsCount";
  ExpressionType2["strConcat"] = "strConcat";
  ExpressionType2["strIndex"] = "strIndex";
  ExpressionType2["strSub"] = "strSub";
  ExpressionType2["viewerSize"] = "viewerSize";
  ExpressionType2["windowSize"] = "windowSize";
  return ExpressionType2;
})(ExpressionType || {});
