// @polygonjs/polygonjs/dist/src/engine/nodes/js/GetMaterial.js
"use strict";
import { TypedJsNode } from "./_Base";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { JsConnectionPoint, JsConnectionPointType } from "../utils/io/connections/Js";
import { NodeContext } from "../../poly/NodeContext";
import { Poly } from "../../Poly";
class GetMaterialJsParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    /** @param the material node */
    this.node = ParamConfig.NODE_PATH("", {
      nodeSelection: {
        context: NodeContext.MAT
      },
      dependentOnFoundNode: false,
      computeOnDirty: true
    });
  }
}
const ParamsConfig = new GetMaterialJsParamsConfig();
export class GetMaterialJsNode extends TypedJsNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return "getMaterial";
  }
  initializeNode() {
    this.io.outputs.setNamedOutputConnectionPoints([
      new JsConnectionPoint(JsConnectionPointType.MATERIAL, JsConnectionPointType.MATERIAL)
    ]);
  }
  setLines(shadersCollectionController) {
    const node = this.pv.node.node();
    if (!(node && node.context() == NodeContext.MAT)) {
      return;
    }
    const nodePath = `'${node.path()}'`;
    const varName = this.jsVarName(JsConnectionPointType.MATERIAL);
    const func = Poly.namedFunctionsRegister.getFunction("getMaterial", this, shadersCollectionController);
    shadersCollectionController.addBodyOrComputed(this, [
      {
        dataType: JsConnectionPointType.MATERIAL,
        varName,
        value: func.asString(nodePath)
      }
    ]);
  }
}
