// @polygonjs/polygonjs/dist/src/engine/nodes/js/Mod.js
"use strict";
import { Poly } from "../../Poly";
import { JsConnectionPointType } from "../utils/io/connections/Js";
import { MathFunctionArg2OperationFactory } from "./_Math_Arg1Operation";
var ModInput = /* @__PURE__ */ ((ModInput2) => {
  ModInput2["value0"] = "value0";
  ModInput2["value1"] = "value1";
  return ModInput2;
})(ModInput || {});
const DefaultValues = {
  ["value0" /* value0 */]: 0,
  ["value1" /* value1 */]: 1
};
const FUNCTION_NAME = "mod";
export class ModJsNode extends MathFunctionArg2OperationFactory("mod", {
  inputPrefix: "in",
  out: "mod"
}) {
  _coreFunction(shadersCollectionController) {
    Poly.namedFunctionsRegister.getFunction(FUNCTION_NAME, this, shadersCollectionController).asString("", "");
    return FUNCTION_NAME;
  }
  paramDefaultValue(name) {
    return DefaultValues[name];
  }
  _expectedInputName(index) {
    return ["value0" /* value0 */, "value1" /* value1 */][index];
  }
  _expectedInputTypes() {
    return [JsConnectionPointType.FLOAT, JsConnectionPointType.FLOAT];
  }
}
