// @polygonjs/polygonjs/dist/src/engine/nodes/js/PhysicsRBDResetTorques.js
"use strict";
import { ParamConfig } from "./../utils/params/ParamsConfig";
import { TRIGGER_CONNECTION_NAME, TypedJsNode } from "./_Base";
import { NodeParamsConfig } from "../utils/params/ParamsConfig";
import { JsConnectionPoint, JsConnectionPointType, JS_CONNECTION_POINT_IN_NODE_DEF } from "../utils/io/connections/Js";
import { inputObject3D } from "./_BaseObject3D";
import { Poly } from "../../Poly";
const CONNECTION_OPTIONS = JS_CONNECTION_POINT_IN_NODE_DEF;
class PhysicsRBDResetTorquesJsParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    /** @param wakeup */
    this.wakeup = ParamConfig.BOOLEAN(0);
  }
}
const ParamsConfig = new PhysicsRBDResetTorquesJsParamsConfig();
export class PhysicsRBDResetTorquesJsNode extends TypedJsNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return "physicsRBDResetTorques";
  }
  initializeNode() {
    this.io.inputs.setNamedInputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER, CONNECTION_OPTIONS),
      new JsConnectionPoint(JsConnectionPointType.OBJECT_3D, JsConnectionPointType.OBJECT_3D, CONNECTION_OPTIONS)
    ]);
    this.io.outputs.setNamedOutputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER),
      new JsConnectionPoint(JsConnectionPointType.OBJECT_3D, JsConnectionPointType.OBJECT_3D, CONNECTION_OPTIONS)
    ]);
  }
  setTriggerableLines(shadersCollectionController) {
    const object3D = inputObject3D(this, shadersCollectionController);
    const wakeup = this.variableForInputParam(shadersCollectionController, this.p.wakeup);
    const func = Poly.namedFunctionsRegister.getFunction(
      "physicsRBDResetTorques",
      this,
      shadersCollectionController
    );
    const bodyLine = func.asString(object3D, wakeup);
    shadersCollectionController.addTriggerableLines(this, [bodyLine]);
  }
}
