// @polygonjs/polygonjs/dist/src/engine/nodes/js/SetPerspectiveCameraNearFar.js
"use strict";
import { TRIGGER_CONNECTION_NAME, TypedJsNode } from "./_Base";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { JsConnectionPoint, JsConnectionPointType, JS_CONNECTION_POINT_IN_NODE_DEF } from "../utils/io/connections/Js";
import { inputObject3D, setObject3DOutputLine } from "./_BaseObject3D";
import { Poly } from "../../Poly";
const CONNECTION_OPTIONS = JS_CONNECTION_POINT_IN_NODE_DEF;
class SetPerspectiveCameraNearFarJsParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    /** @param near */
    this.near = ParamConfig.FLOAT(1);
    /** @param far */
    this.far = ParamConfig.FLOAT(100);
    /** @param lerp factor */
    this.lerp = ParamConfig.FLOAT(1);
    /** @param sets if the projection matrix should be updated as the animation progresses */
    this.updateProjectionMatrix = ParamConfig.BOOLEAN(1);
  }
}
const ParamsConfig = new SetPerspectiveCameraNearFarJsParamsConfig();
export class SetPerspectiveCameraNearFarJsNode extends TypedJsNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return "setPerspectiveCameraNearFar";
  }
  initializeNode() {
    this.io.inputs.setNamedInputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER, CONNECTION_OPTIONS),
      new JsConnectionPoint(JsConnectionPointType.OBJECT_3D, JsConnectionPointType.OBJECT_3D, CONNECTION_OPTIONS)
    ]);
    this.io.outputs.setNamedOutputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER),
      new JsConnectionPoint(JsConnectionPointType.OBJECT_3D, JsConnectionPointType.OBJECT_3D, CONNECTION_OPTIONS)
    ]);
  }
  setLines(linesController) {
    setObject3DOutputLine(this, linesController);
  }
  setTriggerableLines(shadersCollectionController) {
    const object3D = inputObject3D(this, shadersCollectionController);
    const near = this.variableForInputParam(shadersCollectionController, this.p.near);
    const far = this.variableForInputParam(shadersCollectionController, this.p.far);
    const lerp = this.variableForInputParam(shadersCollectionController, this.p.lerp);
    const updateProjectionMatrix = this.variableForInputParam(
      shadersCollectionController,
      this.p.updateProjectionMatrix
    );
    const func = Poly.namedFunctionsRegister.getFunction(
      "setPerspectiveCameraNearFar",
      this,
      shadersCollectionController
    );
    const bodyLine = func.asString(object3D, near, far, lerp, updateProjectionMatrix);
    shadersCollectionController.addTriggerableLines(this, [bodyLine]);
  }
}
