// @polygonjs/polygonjs/dist/src/engine/nodes/js/SetPhysicsRBDLinearVelocity.js
"use strict";
import { ParamConfig } from "./../utils/params/ParamsConfig";
import { TRIGGER_CONNECTION_NAME, TypedJsNode } from "./_Base";
import { NodeParamsConfig } from "../utils/params/ParamsConfig";
import { JsConnectionPoint, JsConnectionPointType, JS_CONNECTION_POINT_IN_NODE_DEF } from "../utils/io/connections/Js";
import { inputObject3D, setObject3DOutputLine } from "./_BaseObject3D";
import { Poly } from "../../Poly";
const CONNECTION_OPTIONS = JS_CONNECTION_POINT_IN_NODE_DEF;
class SetPhysicsRBDLinearVelocityJsParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    /** @param target position */
    this.velocity = ParamConfig.VECTOR3([0, 0, 0]);
    /** @param lerp factor */
    this.lerp = ParamConfig.FLOAT(1);
  }
}
const ParamsConfig = new SetPhysicsRBDLinearVelocityJsParamsConfig();
export class SetPhysicsRBDLinearVelocityJsNode extends TypedJsNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return "setPhysicsRBDLinearVelocity";
  }
  initializeNode() {
    this.io.inputs.setNamedInputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER, CONNECTION_OPTIONS),
      new JsConnectionPoint(JsConnectionPointType.OBJECT_3D, JsConnectionPointType.OBJECT_3D, CONNECTION_OPTIONS)
    ]);
    this.io.outputs.setNamedOutputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER),
      new JsConnectionPoint(JsConnectionPointType.OBJECT_3D, JsConnectionPointType.OBJECT_3D, CONNECTION_OPTIONS)
    ]);
  }
  setLines(linesController) {
    setObject3DOutputLine(this, linesController);
  }
  setTriggerableLines(shadersCollectionController) {
    const object3D = inputObject3D(this, shadersCollectionController);
    const velocity = this.variableForInputParam(shadersCollectionController, this.p.velocity);
    const lerp = this.variableForInputParam(shadersCollectionController, this.p.lerp);
    const func = Poly.namedFunctionsRegister.getFunction(
      "setPhysicsRBDLinearVelocity",
      this,
      shadersCollectionController
    );
    const bodyLine = func.asString(object3D, velocity, lerp);
    shadersCollectionController.addTriggerableLines(this, [bodyLine]);
  }
}
