// @polygonjs/polygonjs/dist/src/engine/nodes/sop/CameraControls.js
"use strict";
import { TypedSopNode } from "./_Base";
import { CameraControlsSopOperation } from "../../operations/sop/CameraControls";
import { HierarchyParamConfigAll, ParamConfig } from "../utils/params/ParamsConfig";
import { CameraSopNodeType, NodeContext } from "../../poly/NodeContext";
class CameraControlsSopParamsConfig extends HierarchyParamConfigAll {
  constructor() {
    super(...arguments);
    /** @param renderer */
    this.node = ParamConfig.NODE_PATH("", {
      nodeSelection: {
        context: NodeContext.EVENT
      },
      dependentOnFoundNode: true
    });
  }
}
const ParamsConfig = new CameraControlsSopParamsConfig();
export class CameraControlsSopNode extends TypedSopNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
    /*
    children
    */
    this._childrenControllerContext = NodeContext.EVENT;
  }
  static type() {
    return CameraSopNodeType.CONTROLS;
  }
  initializeNode() {
    this.io.inputs.setCount(1);
    this.io.inputs.initInputsClonedState(CameraControlsSopOperation.INPUT_CLONED_STATE);
  }
  cook(inputCoreGroups) {
    this._operation = this._operation || new CameraControlsSopOperation(this._scene, this.states, this);
    const coreGroup = this._operation.cook(inputCoreGroups, this.pv);
    this.setCoreGroup(coreGroup);
  }
  createNode(node_class, options) {
    return super.createNode(node_class, options);
  }
  children() {
    return super.children();
  }
  nodesByType(type) {
    return super.nodesByType(type);
  }
}
