// @polygonjs/polygonjs/dist/src/engine/nodes/sop/CameraViewOffset.js
"use strict";
import { TypedSopNode } from "./_Base";
import { CameraViewOffsetSopOperation } from "../../operations/sop/CameraViewOffset";
import { HierarchyParamConfigAll } from "../utils/params/ParamsConfig";
import { CameraSopNodeType } from "../../poly/NodeContext";
import { CoreCameraViewOffsetParamConfig } from "../../../core/camera/CoreCameraViewOffset";
class CameraViewOffsetSopParamsConfig extends CoreCameraViewOffsetParamConfig(HierarchyParamConfigAll) {
}
const ParamsConfig = new CameraViewOffsetSopParamsConfig();
export class CameraViewOffsetSopNode extends TypedSopNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return CameraSopNodeType.VIEW_OFFSET;
  }
  initializeNode() {
    this.io.inputs.setCount(1);
    this.io.inputs.initInputsClonedState(CameraViewOffsetSopOperation.INPUT_CLONED_STATE);
  }
  cook(inputCoreGroups) {
    this._operation = this._operation || new CameraViewOffsetSopOperation(this._scene, this.states, this);
    const coreGroup = this._operation.cook(inputCoreGroups, this.pv);
    this.setCoreGroup(coreGroup);
  }
}
