// @polygonjs/polygonjs/dist/src/engine/nodes/sop/TrackingLandmarksFaceAttributes.js
"use strict";
import { TypedSopNode } from "./_Base";
import { NodeParamsConfig } from "../utils/params/ParamsConfig";
import { InputCloneMode } from "../../poly/InputCloneMode";
import {
  CoreComputerVisionFace,
  CoreComputerVisionFaceParamConfig
} from "../../../core/computerVision/face/CoreComputerVisionFace";
class TrackingLandmarksFaceAttributesSopParamsConfig extends CoreComputerVisionFaceParamConfig(NodeParamsConfig) {
}
const ParamsConfig = new TrackingLandmarksFaceAttributesSopParamsConfig();
export class TrackingLandmarksFaceAttributesSopNode extends TypedSopNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return "trackingLandmarksFaceAttributes";
  }
  initializeNode() {
    this.io.inputs.setCount(1);
    this.io.inputs.initInputsClonedState(InputCloneMode.FROM_NODE);
  }
  cook(inputCoreGroups) {
    const inputCoreGroup = inputCoreGroups[0];
    const objects = inputCoreGroup.threejsObjects();
    for (const object of objects) {
      CoreComputerVisionFace.setAttributes(object, this.pv);
    }
    this.setCoreGroup(inputCoreGroup);
  }
}
