// @polygonjs/polygonjs/dist/src/engine/nodes/sop/WFCTileErrorObject.js
"use strict";
import { TypedSopNode } from "./_Base";
import { NodeParamsConfig } from "../utils/params/ParamsConfig";
import { SopType } from "../../poly/registers/nodes/types/Sop";
import { InputCloneMode } from "../../poly/InputCloneMode";
import { createDefaultErrorTileObject, addErrorTileObjectAttributes } from "../../../core/wfc/WFCDebugTileObjects";
class WFCTileErrorObjectSopParamsConfig extends NodeParamsConfig {
}
const ParamsConfig = new WFCTileErrorObjectSopParamsConfig();
export class WFCTileErrorObjectSopNode extends TypedSopNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return SopType.WFC_TILE_ERROR_OBJECT;
  }
  initializeNode() {
    this.io.inputs.setCount(0, 2);
    this.io.inputs.initInputsClonedState(InputCloneMode.FROM_NODE);
  }
  async cook(inputCoreGroups) {
    const coreGroup0 = inputCoreGroups[0];
    const coreGroup1 = inputCoreGroups[1];
    const objects = coreGroup0 ? coreGroup0.allObjects() : [];
    const errorTileObject = coreGroup1 ? coreGroup1.threejsObjects()[0] : null;
    const tileObject = errorTileObject != null ? errorTileObject : createDefaultErrorTileObject();
    addErrorTileObjectAttributes(tileObject);
    objects.push(tileObject);
    this.setObjects(objects);
  }
}
