// @polygonjs/polygonjs/dist/src/engine/operations/sop/CameraViewerCode.js
"use strict";
import { BaseSopOperation } from "./_Base";
import { InputCloneMode } from "../../../engine/poly/InputCloneMode";
import { ThreejsCoreObject } from "../../../core/geometry/modules/three/ThreejsCoreObject";
import { CameraAttribute } from "../../../core/camera/CoreCamera";
import { CameraSopNodeType } from "../../poly/NodeContext";
import { CoreMask } from "../../../core/geometry/Mask";
const _CameraViewerCodeSopOperation = class extends BaseSopOperation {
  static type() {
    return CameraSopNodeType.VIEWER_CODE;
  }
  cook(inputCoreGroups, params) {
    const coreGroup = inputCoreGroups[0];
    const objects = CoreMask.filterObjects(coreGroup, {
      group: params.group
    });
    if (this._node) {
      _CameraViewerCodeSopOperation.updateObject({ objects, params });
    }
    return coreGroup;
  }
  static updateObject(options) {
    const { objects, params } = options;
    for (let object of objects) {
      ThreejsCoreObject.addAttribute(object, CameraAttribute.VIEWER_ID, params.viewerId);
      ThreejsCoreObject.addAttribute(object, CameraAttribute.VIEWER_HTML, params.html);
    }
  }
};
export let CameraViewerCodeSopOperation = _CameraViewerCodeSopOperation;
CameraViewerCodeSopOperation.DEFAULT_PARAMS = {
  group: "",
  viewerId: "my-viewer",
  // shadowRoot: true,
  html: ""
};
CameraViewerCodeSopOperation.INPUT_CLONED_STATE = InputCloneMode.FROM_NODE;
