// @polygonjs/polygonjs/dist/src/engine/operations/sop/CameraFPS.js
"use strict";
import { BaseSopOperation } from "./_Base";
import { InputCloneMode } from "../../../engine/poly/InputCloneMode";
import { ThreejsCoreObject } from "../../../core/geometry/modules/three/ThreejsCoreObject";
import { CameraAttribute } from "../../../core/camera/CoreCamera";
import { CameraSopNodeType } from "../../poly/NodeContext";
import { CoreMask } from "../../../core/geometry/Mask";
const _CameraFPSSopOperation = class extends BaseSopOperation {
  static type() {
    return CameraSopNodeType.FPS;
  }
  cook(inputCoreGroups, params) {
    const coreGroup = inputCoreGroups[0];
    const objects = CoreMask.filterObjects(coreGroup, {
      group: params.group
    });
    if (this._node) {
      _CameraFPSSopOperation.updateObject({ objects, params });
    }
    return coreGroup;
  }
  static updateObject(options) {
    const { objects, params } = options;
    for (let object of objects) {
      ThreejsCoreObject.addAttribute(object, CameraAttribute.MAX_FPS, params.maxFPS);
      ThreejsCoreObject.addAttribute(object, CameraAttribute.MAX_FPS_DYNAMIC_CHANGE, params.allowDynamicChange);
    }
  }
};
export let CameraFPSSopOperation = _CameraFPSSopOperation;
CameraFPSSopOperation.DEFAULT_PARAMS = {
  group: "",
  maxFPS: 60,
  allowDynamicChange: false
};
CameraFPSSopOperation.INPUT_CLONED_STATE = InputCloneMode.FROM_NODE;
