// @polygonjs/polygonjs/dist/src/engine/nodes/gl/SDF2DStairs.js
"use strict";
import { ThreeToGl } from "../../../../src/core/ThreeToGl";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { GlConnectionPointType, GlConnectionPoint } from "../utils/io/connections/Gl";
import { BaseSDF2DGlNode } from "./_BaseSDF2D";
import { GlType } from "../../poly/registers/nodes/types/Gl";
const OUTPUT_NAME = "float";
class SDF2DStairsGlParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    this.position = ParamConfig.VECTOR2([0, 0], { hidden: true });
    this.center = ParamConfig.VECTOR2([0, 0]);
    this.width = ParamConfig.FLOAT(1);
    this.height = ParamConfig.FLOAT(1);
    this.steps = ParamConfig.FLOAT(5, {
      range: [0, 10],
      rangeLocked: [true, false]
    });
  }
}
const ParamsConfig = new SDF2DStairsGlParamsConfig();
export class SDF2DStairsGlNode extends BaseSDF2DGlNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return GlType.SDF_2D_STAIRS;
  }
  initializeNode() {
    super.initializeNode();
    this.io.outputs.setNamedOutputConnectionPoints([
      new GlConnectionPoint(OUTPUT_NAME, GlConnectionPointType.FLOAT)
    ]);
  }
  setLines(shadersCollectionController) {
    const position = this.position();
    const center = ThreeToGl.vector3(this.variableForInputParam(this.p.center));
    const width = ThreeToGl.float(this.variableForInputParam(this.p.width));
    const height = ThreeToGl.float(this.variableForInputParam(this.p.height));
    const steps = ThreeToGl.float(this.variableForInputParam(this.p.steps));
    const float = this.glVarName(OUTPUT_NAME);
    const bodyLine = `float ${float} = sdStairs(${position} - ${center}, vec2(${width}, ${height}), ${steps})`;
    shadersCollectionController.addBodyLines(this, [bodyLine]);
    this._addSDF2DMethods(shadersCollectionController);
  }
}
