// @polygonjs/polygonjs/dist/src/engine/nodes/gl/SDFOctogonalPrism.js
"use strict";
import { ThreeToGl } from "../../../core/ThreeToGl";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { GlConnectionPointType, GlConnectionPoint } from "../utils/io/connections/Gl";
import { BaseSDFGlNode } from "./_BaseSDF";
import { GlType } from "../../poly/registers/nodes/types/Gl";
const OUTPUT_NAME = "float";
class SDFOctogonalPrismGlParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    this.position = ParamConfig.VECTOR3([0, 0, 0], { hidden: true });
    this.center = ParamConfig.VECTOR3([0, 0, 0]);
    this.radius = ParamConfig.FLOAT(0.1);
    this.height = ParamConfig.FLOAT(1);
  }
}
const ParamsConfig = new SDFOctogonalPrismGlParamsConfig();
export class SDFOctogonalPrismGlNode extends BaseSDFGlNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return GlType.SDF_OCTOGONAL_PRISM;
  }
  initializeNode() {
    super.initializeNode();
    this.io.outputs.setNamedOutputConnectionPoints([
      new GlConnectionPoint(OUTPUT_NAME, GlConnectionPointType.FLOAT)
    ]);
  }
  setLines(shadersCollectionController) {
    const position = this.position();
    const center = ThreeToGl.vector2(this.variableForInputParam(this.p.center));
    const radius = ThreeToGl.float(this.variableForInputParam(this.p.radius));
    const height = ThreeToGl.float(this.variableForInputParam(this.p.height));
    const float = this.glVarName(OUTPUT_NAME);
    const bodyLine = `float ${float} = sdOctogonPrism(${position} - ${center}, ${radius}, ${height})`;
    shadersCollectionController.addBodyLines(this, [bodyLine]);
    this._addSDFMethods(shadersCollectionController);
  }
}
