// @polygonjs/polygonjs/dist/src/engine/nodes/gl/UvToOklab.js
"use strict";
import { TypedGlNode } from "./_Base";
import { ThreeToGl } from "../../../../src/core/ThreeToGl";
import oklab from "./gl/oklab.glsl";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { FunctionGLDefinition } from "./utils/GLDefinition";
import { GlConnectionPoint, GlConnectionPointType } from "../utils/io/connections/Gl";
const OUTPUT_NAME = "oklab";
class UvToOklabGlParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    this.uvw = ParamConfig.VECTOR3([1, 1, 1]);
  }
}
const ParamsConfig = new UvToOklabGlParamsConfig();
export class UvToOklabGlNode extends TypedGlNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return "uvToOklab";
  }
  initializeNode() {
    this.io.outputs.setNamedOutputConnectionPoints([
      new GlConnectionPoint(OUTPUT_NAME, GlConnectionPointType.VEC3)
    ]);
  }
  setLines(shaders_collection_controller) {
    const function_declaration_lines = [];
    const body_lines = [];
    function_declaration_lines.push(new FunctionGLDefinition(this, oklab));
    const uvw = ThreeToGl.vector3(this.variableForInputParam(this.p.uvw));
    const oklabOut = this.glVarName(OUTPUT_NAME);
    body_lines.push(`vec3 ${oklabOut} = uvToOklab(${uvw})`);
    shaders_collection_controller.addDefinitions(this, function_declaration_lines);
    shaders_collection_controller.addBodyLines(this, body_lines);
  }
}
