// @polygonjs/polygonjs/dist/src/engine/nodes/js/code/assemblers/objectBuilder/ObjectBuilderAssemblerCommon.js
"use strict";
export var ObjectBuilderAssemblerConstant = /* @__PURE__ */ ((ObjectBuilderAssemblerConstant2) => {
  ObjectBuilderAssemblerConstant2["OBJECT_CONTAINER"] = "objectContainer";
  ObjectBuilderAssemblerConstant2["OBJECT_3D"] = "objectContainer.Object3D";
  ObjectBuilderAssemblerConstant2["OBJ_NUM"] = "objectContainer.objnum";
  ObjectBuilderAssemblerConstant2["MATERIAL"] = "objectContainer.Object3D.material";
  ObjectBuilderAssemblerConstant2["PTNUM"] = "null";
  ObjectBuilderAssemblerConstant2["PRIMITIVE_GRAPH"] = "null";
  return ObjectBuilderAssemblerConstant2;
})(ObjectBuilderAssemblerConstant || {});
export var ObjectVariable = /* @__PURE__ */ ((ObjectVariable2) => {
  ObjectVariable2["OBJECT_3D"] = "Object3D";
  ObjectVariable2["POSITION"] = "position";
  ObjectVariable2["ROTATION"] = "rotation";
  ObjectVariable2["QUATERNION"] = "quaternion";
  ObjectVariable2["SCALE"] = "scale";
  ObjectVariable2["MATRIX"] = "matrix";
  ObjectVariable2["VISIBLE"] = "visible";
  ObjectVariable2["MATRIX_AUTO_UPDATE"] = "matrixAutoUpdate";
  ObjectVariable2["CAST_SHADOW"] = "castShadow";
  ObjectVariable2["RECEIVE_SHADOW"] = "receiveShadow";
  ObjectVariable2["FRUSTUM_CULLED"] = "frustumCulled";
  ObjectVariable2["OBJ_NUM"] = "objnum";
  return ObjectVariable2;
})(ObjectVariable || {});
