// @polygonjs/polygonjs/dist/src/engine/nodes/obj/utils/TesselationParams.js
"use strict";
import { TypedObjNode } from "../_Base";
import { NodeParamsConfig, ParamConfig } from "../../utils/params/ParamsConfig";
import { OBJCADTesselationParamConfig } from "../../../../core/geometry/modules/cad/utils/TesselationParamsConfig";
import { OBJCSGTesselationParamConfig } from "../../../../core/geometry/modules/csg/utils/TesselationParamsConfig";
import { OBJQUADTesselationParamConfig } from "../../../../core/geometry/modules/quad/utils/TesselationParamsConfig";
import { OBJTetTesselationParamConfig } from "../../../../core/geometry/modules/tet/utils/TesselationParamsConfig";
export function ObjCADTesselationFolderParamConfig(Base) {
  return class Mixin extends Base {
    constructor() {
      super(...arguments);
      this.CAD = ParamConfig.FOLDER();
    }
  };
}
export function ObjCSGTesselationFolderParamConfig(Base) {
  return class Mixin extends Base {
    constructor() {
      super(...arguments);
      this.CSG = ParamConfig.FOLDER();
    }
  };
}
export function ObjQUADTesselationFolderParamConfig(Base) {
  return class Mixin extends Base {
    constructor() {
      super(...arguments);
      this.QUAD = ParamConfig.FOLDER();
    }
  };
}
export function ObjTetTesselationFolderParamConfig(Base) {
  return class Mixin extends Base {
    constructor() {
      super(...arguments);
      this.TET = ParamConfig.FOLDER();
    }
  };
}
export function ObjTesselationParamConfig(Base) {
  return class Mixin extends OBJTetTesselationParamConfig(
    ObjTetTesselationFolderParamConfig(
      OBJQUADTesselationParamConfig(
        ObjQUADTesselationFolderParamConfig(
          OBJCSGTesselationParamConfig(
            ObjCSGTesselationFolderParamConfig(
              OBJCADTesselationParamConfig(ObjCADTesselationFolderParamConfig(Base))
            )
          )
        )
        //		)
        //)
      )
    )
  ) {
  };
}
class TesselationParamParamsConfig extends ObjTesselationParamConfig(NodeParamsConfig) {
}
export class TesselationParamsObjNode extends TypedObjNode {
}
