// @polygonjs/polygonjs/dist/src/core/MouseButton.js
"use strict";
export var MouseButton = /* @__PURE__ */ ((MouseButton2) => {
  MouseButton2[MouseButton2["LEFT"] = 0] = "LEFT";
  MouseButton2[MouseButton2["MIDDLE"] = 1] = "MIDDLE";
  MouseButton2[MouseButton2["RIGHT"] = 2] = "RIGHT";
  return MouseButton2;
})(MouseButton || {});
export var MouseButtons = /* @__PURE__ */ ((MouseButtons2) => {
  MouseButtons2[MouseButtons2["NONE"] = 0] = "NONE";
  MouseButtons2[MouseButtons2["LEFT"] = 1] = "LEFT";
  MouseButtons2[MouseButtons2["RIGHT"] = 2] = "RIGHT";
  MouseButtons2[MouseButtons2["LEFT_RIGHT"] = 3] = "LEFT_RIGHT";
  MouseButtons2[MouseButtons2["MIDDLE"] = 4] = "MIDDLE";
  MouseButtons2[MouseButtons2["LEFT_MIDDLE"] = 5] = "LEFT_MIDDLE";
  MouseButtons2[MouseButtons2["MIDDLE_RIGHT"] = 6] = "MIDDLE_RIGHT";
  MouseButtons2[MouseButtons2["LEFT_MIDDLE_RIGHT"] = 7] = "LEFT_MIDDLE_RIGHT";
  return MouseButtons2;
})(MouseButtons || {});
export const MOUSE_BUTTONS = [
  /*0*/
  [],
  /*1*/
  [0 /* LEFT */],
  /*2*/
  [2 /* RIGHT */],
  /*3*/
  [0 /* LEFT */, 2 /* RIGHT */],
  /*4*/
  [1 /* MIDDLE */],
  /*5*/
  [0 /* LEFT */, 1 /* MIDDLE */],
  /*6*/
  [1 /* MIDDLE */, 2 /* RIGHT */],
  /*7*/
  [0 /* LEFT */, 1 /* MIDDLE */, 2 /* RIGHT */]
];
