// @polygonjs/polygonjs/dist/src/engine/poly/registers/nodes/types/Mat.js
"use strict";
export var MatType = /* @__PURE__ */ ((MatType2) => {
  MatType2["LINE_BASIC"] = "lineBasic";
  MatType2["LINE_BASIC_BUILDER"] = "lineBasicBuilder";
  MatType2["MESH_BASIC"] = "meshBasic";
  MatType2["MESH_BASIC_BUILDER"] = "meshBasicBuilder";
  MatType2["MESH_DEPTH_BUILDER"] = "meshDepthBuilder";
  MatType2["MESH_DISTANCE_BUILDER"] = "meshDistanceBuilder";
  MatType2["MESH_LAMBERT"] = "meshLambert";
  MatType2["MESH_LAMBERT_BUILDER"] = "meshLambertBuilder";
  MatType2["MESH_MATCAP"] = "meshMatcap";
  MatType2["MESH_NORMAL"] = "meshNormal";
  MatType2["MESH_PHONG"] = "meshPhong";
  MatType2["MESH_PHONG_BUILDER"] = "meshPhongBuilder";
  MatType2["MESH_PHYSICAL"] = "meshPhysical";
  MatType2["MESH_PHYSICAL_BUILDER"] = "meshPhysicalBuilder";
  MatType2["MESH_STANDARD"] = "meshStandard";
  MatType2["MESH_STANDARD_BUILDER"] = "meshStandardBuilder";
  MatType2["MESH_TOON"] = "meshToon";
  MatType2["MESH_TOON_BUILDER"] = "meshToonBuilder";
  MatType2["POINTS"] = "points";
  MatType2["POINTS_BUILDER"] = "pointsBuilder";
  MatType2["RAY_MARCHING_BUILDER"] = "rayMarchingBuilder";
  MatType2["SHADOW"] = "shadow";
  MatType2["SKY"] = "sky";
  MatType2["VOLUME"] = "volume";
  MatType2["VOLUME_BUILDER"] = "volumeBuilder";
  return MatType2;
})(MatType || {});
