// @polygonjs/polygonjs/dist/src/engine/poly/registers/nodes/types/Sop.js
"use strict";
export var SopType = /* @__PURE__ */ ((SopType2) => {
  SopType2["ACTOR"] = "actor";
  SopType2["ACTOR_INSTANCE"] = "actorInstance";
  SopType2["ACTOR_POINT"] = "actorPoint";
  SopType2["ADD"] = "add";
  SopType2["ADJACENCY"] = "adjacency";
  SopType2["ANIMATION_COPY"] = "animationCopy";
  SopType2["ATTRIB_ADD_MULT"] = "attribAddMult";
  SopType2["ATTRIB_COPY"] = "attribCopy";
  SopType2["ATTRIB_CREATE"] = "attribCreate";
  SopType2["ATTRIB_DELETE"] = "attribDelete";
  SopType2["ATTRIB_ID"] = "attribId";
  SopType2["ATTRIB_PROMOTE"] = "attribPromote";
  SopType2["ATTRIB_RENAME"] = "attribRename";
  SopType2["ATTRIB_SET_AT_INDEX"] = "attribSetAtIndex";
  SopType2["ATTRIB_TRANSFER"] = "attribTransfer";
  SopType2["AUDIO_NOTES"] = "audioNotes";
  SopType2["AXES_HELPER"] = "axesHelper";
  SopType2["BBOX_SCATTER"] = "bboxScatter";
  SopType2["BLEND"] = "blend";
  SopType2["BOX"] = "box";
  SopType2["BOX_LINES"] = "boxLines";
  SopType2["BVH"] = "BVH";
  SopType2["BVH_VISUALIZER"] = "BVHVisualizer";
  SopType2["CACHE"] = "cache";
  SopType2["CAD_BOOLEAN"] = "CADBoolean";
  SopType2["CAD_BOX"] = "CADBox";
  SopType2["CAD_CIRCLE"] = "CADCircle";
  SopType2["CAD_CIRCLE_2D"] = "CADCircle2D";
  SopType2["CAD_CIRCLE_3_POINTS"] = "CADCircle3Points";
  SopType2["CAD_CONE"] = "CADCone";
  SopType2["CAD_CONVERT_DIMENSION"] = "CADConvertDimension";
  SopType2["CAD_CURVE_2D_TO_SURFACE"] = "CADCurve2DToSurface";
  SopType2["CAD_CURVE_FROM_POINTS"] = "CADCurveFromPoints";
  SopType2["CAD_CURVE_FROM_POINTS_2D"] = "CADCurveFromPoints2D";
  SopType2["CAD_CURVE_TRIM"] = "CADCurveTrim";
  SopType2["CAD_ELLIPSE"] = "CADEllipse";
  SopType2["CAD_ELLIPSE_2D"] = "CADEllipse2D";
  SopType2["CAD_EXPORTER_STEP"] = "CADExporterSTEP";
  SopType2["CAD_EXTRUDE"] = "CADExtrude";
  SopType2["CAD_FILE_STEP"] = "CADFileSTEP";
  SopType2["CAD_FILLET"] = "CADFillet";
  SopType2["CAD_GROUP"] = "CADGroup";
  SopType2["CAD_LOFT"] = "CADLoft";
  SopType2["CAD_MIRROR"] = "CADMirror";
  SopType2["CAD_PIPE"] = "CADPipe";
  SopType2["CAD_POINT"] = "CADPoint";
  SopType2["CAD_POINT_2D"] = "CADPoint2D";
  SopType2["CAD_POINTS_FROM_CURVE"] = "CADPointsFromCurve";
  SopType2["CAD_RECTANGLE"] = "CADRectangle";
  SopType2["CAD_REVOLUTION"] = "CADRevolution";
  SopType2["CAD_SEGMENT"] = "CADSegment";
  SopType2["CAD_SPHERE"] = "CADSphere";
  SopType2["CAD_THICKNESS"] = "CADThickness";
  SopType2["CAD_TORUS"] = "CADTorus";
  SopType2["CAD_TRANSFORM"] = "CADTransform";
  SopType2["CAD_TRANSFORM_2D"] = "CADTransform2D";
  SopType2["CAD_TRIANGULATE"] = "CADTriangulate";
  SopType2["CAD_TUBE"] = "CADTube";
  SopType2["CAD_UNPACK"] = "CADUnpack";
  SopType2["CAD_WEDGE"] = "CADWedge";
  SopType2["CAMERA_PROJECT"] = "cameraProject";
  SopType2["CAPSULE"] = "capsule";
  SopType2["CIRCLE"] = "circle";
  SopType2["CIRCLE_3_POINTS"] = "circle3Points";
  SopType2["CLOTH_PREPARE"] = "clothPrepare";
  SopType2["CLOTH_SOLVER"] = "clothSolver";
  SopType2["COLOR"] = "color";
  SopType2["CONE"] = "cone";
  SopType2["CONTACT_SHADOWS"] = "contactShadows";
  SopType2["COPY"] = "copy";
  SopType2["CURVE_FROM_POINTS"] = "curveFromPoints";
  SopType2["CSG_ARC"] = "CSGArc";
  SopType2["CSG_BOOLEAN"] = "CSGBoolean";
  SopType2["CSG_CENTER"] = "CSGCenter";
  SopType2["CSG_CIRCLE"] = "CSGCircle";
  SopType2["CSG_BOX"] = "CSGBox";
  SopType2["CSG_DODECAHEDRON"] = "CSGDodecahedron";
  SopType2["CSG_ELLIPSE"] = "CSGEllipse";
  SopType2["CSG_ELLIPSOID"] = "CSGEllipsoid";
  SopType2["CSG_EXPAND"] = "CSGExpand";
  SopType2["CSG_EXTRUDE_LINEAR"] = "CSGExtrudeLinear";
  SopType2["CSG_EXTRUDE_RECTANGULAR"] = "CSGExtrudeRectangular";
  SopType2["CSG_EXTRUDE_ROTATE"] = "CSGExtrudeRotate";
  SopType2["CSG_HULL"] = "CSGHull";
  SopType2["CSG_LINE"] = "CSGLine";
  SopType2["CSG_MIRROR"] = "CSGMirror";
  SopType2["CSG_OFFSET"] = "CSGOffset";
  SopType2["CSG_POLYGON"] = "CSGPolygon";
  SopType2["CSG_POLYHEDRON"] = "CSGPolyhedron";
  SopType2["CSG_PROJECT"] = "CSGProject";
  SopType2["CSG_RECTANGLE"] = "CSGRectangle";
  SopType2["CSG_SPHERE"] = "CSGSphere";
  SopType2["CSG_STAR"] = "CSGStar";
  SopType2["CSG_TORUS"] = "CSGTorus";
  SopType2["CSG_TRANSFORM_2D"] = "CSGTransform2D";
  SopType2["CSG_TRANSFORM_RESET"] = "CSGTransformReset";
  SopType2["CSG_TRIANGLE"] = "CSGTriangle";
  SopType2["CSG_TRIANGULATE"] = "CSGTriangulate";
  SopType2["CSG_TUBE"] = "CSGTube";
  SopType2["CSG_TUBE_ELLIPTIC"] = "CSGTubeElliptic";
  SopType2["CSS2D_OBJECT"] = "CSS2DObject";
  SopType2["CSS3D_OBJECT"] = "CSS3DObject";
  SopType2["DATA"] = "data";
  SopType2["DATA_URL"] = "dataUrl";
  SopType2["DECAL"] = "decal";
  SopType2["DELETE"] = "delete";
  SopType2["DELETE_BY_NAME"] = "deleteByName";
  SopType2["ENTITY_BUILDER"] = "entityBuilder";
  SopType2["FACET"] = "facet";
  SopType2["FUSE"] = "fuse";
  SopType2["GROUND_PROJECTED_SKYBOX"] = "groundProjectedSkybox";
  SopType2["HIERARCHY"] = "hierarchy";
  SopType2["ICOSAHEDRON"] = "icosahedron";
  SopType2["IFC_FILTER_CATEGORIES"] = "IFCFilterCategories";
  SopType2["INSTANCE"] = "instance";
  SopType2["INSTANCE_BUILDER"] = "instanceBuilder";
  SopType2["INSTANCED_MESH_TO_MESH"] = "instancedMeshToMesh";
  SopType2["INSTANCE_UPDATE"] = "instanceUpdate";
  SopType2["JITTER"] = "jitter";
  SopType2["LATTICE"] = "lattice";
  SopType2["LAYER"] = "layer";
  SopType2["LOD"] = "lod";
  SopType2["LOOK_AT"] = "lookAt";
  SopType2["MAPBOX_TRANSFORM"] = "mapboxTransform";
  SopType2["MATERIAL"] = "material";
  SopType2["MATERIAL_PROPERTIES"] = "materialProperties";
  SopType2["MERGE"] = "merge";
  SopType2["METABALL"] = "metaball";
  SopType2["MIRROR"] = "mirror";
  SopType2["NOISE"] = "noise";
  SopType2["NORMALS"] = "normals";
  SopType2["NORMALS_HELPER"] = "normalsHelper";
  SopType2["OBJECT_BUILDER"] = "objectBuilder";
  SopType2["OBJECT_PROPERTIES"] = "objectProperties";
  SopType2["OBJECTS_LAYOUT"] = "objectsLayout";
  SopType2["PALETTE"] = "palette";
  SopType2["PARTICLES_SYSTEM_GPU"] = "particlesSystemGpu";
  SopType2["PARTICLES_SYSTEM_GPU_ATTRIBUTES"] = "particlesSystemGpuAttributes";
  SopType2["PARTICLES_SYSTEM_GPU_MATERIAL"] = "particlesSystemGpuMaterial";
  SopType2["PHYSICS_DEBUG"] = "physicsDebug";
  SopType2["PHYSICS_GROUND"] = "physicsGround";
  SopType2["PHYSICS_RBD_ATTRIBUTES"] = "physicsRBDAttributes";
  SopType2["PHYSICS_RBD_JOINTS"] = "physicsRBDJoints";
  SopType2["PHYSICS_WORLD"] = "physicsWorld";
  SopType2["PHYSICS_PLAYER"] = "physicsPlayer";
  SopType2["PLANE"] = "plane";
  SopType2["POINT"] = "point";
  SopType2["POINT_BUILDER"] = "pointBuilder";
  SopType2["POLAR_TRANSFORM"] = "polarTransform";
  SopType2["POLYWIRE"] = "polywire";
  SopType2["QUAD_CORNERS"] = "quadCorners";
  SopType2["QUAD_EXTRUDE"] = "quadExtrude";
  SopType2["QUAD_PLANE"] = "quadPlane";
  SopType2["QUAD_SMOOTH"] = "quadSmooth";
  SopType2["QUAD_TRIANGULATE"] = "quadTriangulate";
  SopType2["QUAD_UNIQUE_NEIGHBOUR_ID"] = "quadUniqueNeighbourId";
  SopType2["QUADRANGULATE"] = "quadrangulate";
  SopType2["RAY"] = "ray";
  SopType2["REFLECTOR"] = "reflector";
  SopType2["RING"] = "ring";
  SopType2["ROUNDED_BOX"] = "roundedBox";
  SopType2["SCATTER"] = "scatter";
  SopType2["SDF_BUILDER"] = "SDFBuilder";
  SopType2["SET_CHILDREN"] = "setChildren";
  SopType2["SET_GEOMETRY"] = "setGeometry";
  SopType2["SHEAR"] = "shear";
  SopType2["SHORTEST_PATH"] = "shortestPath";
  SopType2["SKELETON_HELPER"] = "skeletonHelper";
  SopType2["SKIN"] = "skin";
  SopType2["SORT"] = "sort";
  SopType2["SPHERE"] = "sphere";
  SopType2["SPLIT"] = "split";
  SopType2["SUBDIVIDE"] = "subdivide";
  SopType2["SWITCH"] = "switch";
  SopType2["TANGENT"] = "tangent";
  SopType2["TANGENTS_HELPER"] = "tangentsHelper";
  SopType2["TET_DELETE"] = "tetDelete";
  SopType2["TET_EMBED"] = "tetEmbed";
  SopType2["TET_SOFT_BODY_SOLVER"] = "tetSoftBodySolver";
  SopType2["TET_TRIANGULATE"] = "tetTriangulate";
  SopType2["TETRAHEDRALIZE"] = "tetrahedralize";
  SopType2["TETRAHEDRON"] = "tetrahedron";
  SopType2["TEXT"] = "text";
  SopType2["TEXTURE_COPY"] = "textureCopy";
  SopType2["TEXTURE_PROPERTIES"] = "textureProperties";
  SopType2["TORUS"] = "torus";
  SopType2["TORUS_KNOT"] = "torusKnot";
  SopType2["TRANSFORM"] = "transform";
  SopType2["TRANSFORM_COPY"] = "transformCopy";
  SopType2["TRANSFORM_MULTI"] = "transformMulti";
  SopType2["TRANSFORM_RESET"] = "transformReset";
  SopType2["TUBE"] = "tube";
  SopType2["UV_LAYOUT"] = "uvLayout";
  SopType2["UV_TRANSFORM"] = "uvTransform";
  SopType2["UV_UNWRAP"] = "uvUnwrap";
  SopType2["WEB_XR_AR_ESTIMATED_LIGHT"] = "webXRAREstimatedLight";
  SopType2["WFC_BUILDER"] = "WFCBuilder";
  SopType2["WFC_DEBUG"] = "WFCDebug";
  SopType2["WFC_SOLVER"] = "WFCSolver";
  SopType2["WFC_RULE_FROM_PROXIMITY"] = "WFCRuleFromProximity";
  SopType2["WFC_RULE_CONNECTION_FROM_SIDE_NAME"] = "WFCRuleConnectionFromSideName";
  SopType2["WFC_RULE_CONNECTION_TO_GRID_BORDER"] = "WFCRuleConnectionToGridBorder";
  SopType2["WFC_RULE_TILE_WEIGHT"] = "WFCRuleTileWeight";
  SopType2["WFC_TILE_EMPTY_OBJECT"] = "WFCTileEmptyObject";
  SopType2["WFC_TILE_ERROR_OBJECT"] = "WFCTileErrorObject";
  SopType2["WFC_TILE_UNRESOLVED_OBJECT"] = "WFCTileUnresolvedObject";
  SopType2["WFC_TILE_PROPERTIES"] = "WFCTileProperties";
  SopType2["WFC_TILE_SIDE_NAME"] = "WFCTileSideName";
  SopType2["WFC_TILE_TRANSFORM"] = "WFCTileTransform";
  return SopType2;
})(SopType || {});
export var SopTypeFile = /* @__PURE__ */ ((SopTypeFile2) => {
  SopTypeFile2["FILE_3DS"] = "file3DS";
  SopTypeFile2["FILE_DRC"] = "fileDRC";
  SopTypeFile2["FILE_FBX"] = "fileFBX";
  SopTypeFile2["FILE_GEOJSON"] = "fileGEOJSON";
  SopTypeFile2["FILE_GLTF"] = "fileGLTF";
  SopTypeFile2["FILE_JSON"] = "fileJSON";
  SopTypeFile2["FILE_MPD"] = "fileMPD";
  SopTypeFile2["FILE_OBJ"] = "fileOBJ";
  SopTypeFile2["FILE_PDB"] = "filePDB";
  SopTypeFile2["FILE_PLY"] = "filePLY";
  SopTypeFile2["FILE_STL"] = "fileSTL";
  SopTypeFile2["FILE_SVG"] = "fileSVG";
  SopTypeFile2["FILE_USDZ"] = "fileUSDZ";
  SopTypeFile2["FILE_VOX"] = "fileVOX";
  return SopTypeFile2;
})(SopTypeFile || {});
export var SopTypeFileMulti = /* @__PURE__ */ ((SopTypeFileMulti2) => {
  SopTypeFileMulti2["FILE_DRC"] = "fileMultiDRC";
  SopTypeFileMulti2["FILE_FBX"] = "fileMultiFBX";
  SopTypeFileMulti2["FILE_GLTF"] = "fileMultiGLTF";
  SopTypeFileMulti2["FILE_JSON"] = "fileMultiJSON";
  SopTypeFileMulti2["FILE_MPD"] = "fileMultiMPD";
  SopTypeFileMulti2["FILE_OBJ"] = "fileMultiOBJ";
  SopTypeFileMulti2["FILE_PDB"] = "fileMultiPDB";
  SopTypeFileMulti2["FILE_PLY"] = "fileMultiPLY";
  SopTypeFileMulti2["FILE_STL"] = "fileMultiSTL";
  SopTypeFileMulti2["FILE_SVG"] = "fileMultiSVG";
  return SopTypeFileMulti2;
})(SopTypeFileMulti || {});
export var SopExporter = /* @__PURE__ */ ((SopExporter2) => {
  SopExporter2["EXPORTER_GLTF"] = "exporterGLTF";
  SopExporter2["EXPORTER_OBJ"] = "exporterOBJ";
  SopExporter2["EXPORTER_PLY"] = "exporterPLY";
  SopExporter2["EXPORTER_STL"] = "exporterSTL";
  SopExporter2["EXPORTER_USDZ"] = "exporterUSDZ";
  return SopExporter2;
})(SopExporter || {});
