// @polygonjs/polygonjs/dist/src/engine/params/Vector2.js
"use strict";
import { TypedMultipleParam } from "./_Multiple";
import { Vector2 } from "three";
import { ParamType } from "../poly/ParamType";
import { CoreType } from "../../core/Type";
const COMPONENT_NAMES_VECTOR2 = ["x", "y"];
const tmp = [0, 0];
function vector2EqualsStringNumber2(v, array) {
  return v.x == array[0] && v.y == array[1];
}
function stringNumber2Equals(array1, array2) {
  return array1[0] == array2[0] && array1[1] == array2[1];
}
export class Vector2Param extends TypedMultipleParam {
  constructor() {
    super(...arguments);
    this._value = new Vector2();
    this._copied_value = [0, 0];
  }
  static type() {
    return ParamType.VECTOR2;
  }
  componentNames() {
    return COMPONENT_NAMES_VECTOR2;
  }
  defaultValueSerialized() {
    if (CoreType.isArray(this._default_value)) {
      return this._default_value;
    } else {
      return this._default_value.toArray();
    }
  }
  _prefilterInvalidRawInput(rawInput) {
    if (rawInput instanceof Vector2) {
      rawInput.toArray(tmp);
      return tmp;
    }
    return super._prefilterInvalidRawInput(rawInput);
  }
  // get raw_input_serialized() {
  // 	if (this._raw_input instanceof Vector2) {
  // 		return this._raw_input.toArray() as Number2;
  // 	} else {
  // 		const new_array: StringOrNumber2 = [this._raw_input[0], this._raw_input[1]];
  // 		return new_array;
  // 	}
  // }
  valueSerialized() {
    return this.value.toArray();
  }
  _copyValue(param) {
    param.value.toArray(this._copied_value);
    this.set(this._copied_value);
  }
  _cloneRawInput(raw_input) {
    if (raw_input instanceof Vector2) {
      return raw_input.clone();
    } else {
      const new_array = [raw_input[0], raw_input[1]];
      if (new_array[0] == null) {
        new_array[0] = new_array[0] || 0;
      }
      if (new_array[1] == null) {
        new_array[1] = new_array[1] || new_array[0];
      }
      return new_array;
    }
  }
  static areRawInputEqual(raw_input1, raw_input2) {
    if (raw_input1 instanceof Vector2) {
      if (raw_input2 instanceof Vector2) {
        return raw_input1.equals(raw_input2);
      } else {
        return vector2EqualsStringNumber2(raw_input1, raw_input2);
      }
    } else {
      if (raw_input2 instanceof Vector2) {
        return vector2EqualsStringNumber2(raw_input2, raw_input1);
      } else {
        return stringNumber2Equals(raw_input1, raw_input2);
      }
    }
  }
  static areValuesEqual(val1, val2) {
    return val1.equals(val2);
  }
  initComponents() {
    super.initComponents();
    this.x = this.components[0];
    this.y = this.components[1];
  }
  setValueFromComponents() {
    this._value.x = this.x.value;
    this._value.y = this.y.value;
  }
  // convert(input: any) {
  // 	if (CoreType.isArray(input)) {
  // 		return new Vector2().fromArray(input);
  // 	}
  // 	return new Vector2();
  // }
}
