// @polygonjs/polygonjs/dist/src/engine/nodes/audio/PingPongDelay.js
"use strict";
import { PingPongDelay } from "tone/build/esm/effect/PingPongDelay";
const DEFAULTS = {
  delayTime: 0.25,
  // feedback: 0.5,
  maxDelay: 1
};
import { TypedAudioNode } from "./_Base";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
class PingPongDelayAudioParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    /** @param delayTime */
    this.delayTime = ParamConfig.FLOAT(DEFAULTS.delayTime, {
      range: [0, 1],
      rangeLocked: [true, false]
    });
    /** @param maxDelay */
    this.maxDelay = ParamConfig.FLOAT(DEFAULTS.maxDelay, {
      range: [0, 1],
      rangeLocked: [true, false]
    });
  }
}
const ParamsConfig = new PingPongDelayAudioParamsConfig();
export class PingPongDelayAudioNode extends TypedAudioNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return "pingPongDelay";
  }
  initializeNode() {
    this.io.inputs.setCount(1);
  }
  cook(inputContents) {
    const audioBuilder = inputContents[0];
    const effect = this._createEffect();
    const inputNode = audioBuilder.audioNode();
    if (inputNode) {
      inputNode.connect(effect);
    }
    audioBuilder.setAudioNode(effect);
    this.setAudioBuilder(audioBuilder);
  }
  _createEffect() {
    return new PingPongDelay({
      delayTime: this.pv.delayTime,
      maxDelay: this.pv.maxDelay
    });
  }
}
