// @polygonjs/polygonjs/dist/src/core/geometry/AttributeUtils.js
"use strict";
import { BaseCoreObject } from "../geometry/entities/object/BaseCoreObject";
export function setObjectVector4(object, attribName, value) {
  BaseCoreObject.addAttribute(object, attribName, value);
}
export function getObjectVector4(object, attribName, target) {
  BaseCoreObject.attribValue(object, attribName, 0, target);
}
export function setObjectVector3(object, attribName, value) {
  BaseCoreObject.addAttribute(object, attribName, value);
}
export function getObjectVector3(object, attribName, target) {
  BaseCoreObject.attribValue(object, attribName, 0, target);
}
export function setObjectVector2(object, attribName, value) {
  BaseCoreObject.addAttribute(object, attribName, value);
}
export function getObjectVector2(object, attribName, target) {
  BaseCoreObject.attribValue(object, attribName, 0, target);
}
export function setObjectNumber(object, attribName, value) {
  BaseCoreObject.addAttribute(object, attribName, value);
}
export function getObjectNumber(object, attribName, defaultValue) {
  const val = BaseCoreObject.attribValue(object, attribName, 0);
  if (val == null) {
    return defaultValue;
  }
  return val;
}
export function setObjectBoolean(object, attribName, value) {
  BaseCoreObject.addAttribute(object, attribName, value);
}
export function getObjectBoolean(object, attribName, defaultValue) {
  const val = BaseCoreObject.attribValue(object, attribName, 0);
  if (val == null) {
    return defaultValue;
  }
  return val;
}
export function setObjectString(object, attribName, value) {
  BaseCoreObject.addAttribute(object, attribName, value);
}
export function getObjectString(object, attribName) {
  return BaseCoreObject.attribValue(object, attribName, 0);
}
