// @polygonjs/polygonjs/dist/src/core/geometry/modules/csg/CsgPoint.js
"use strict";
import { CorePoint } from "../../entities/point/CorePoint";
export class CsgPoint extends CorePoint {
  constructor(object, index) {
    super(object, index);
    this._object = object;
    this._updateGeometry();
  }
  object() {
    return this._object;
  }
  setIndex(index, object) {
    this._index = index;
    if (object) {
      this._object = object;
      this._updateGeometry();
    }
    return this;
  }
  _updateGeometry() {
    const geometry = this._object.geometry;
    if (geometry) {
      this._geometry = geometry;
    }
  }
  geometry() {
    return this._geometry;
  }
  static addAttribute(object, attribName, attribute) {
  }
  static attributes(object) {
    return void 0;
  }
  static entitiesCount(object) {
    return 0;
  }
  position(target) {
    return target;
  }
  normal(target) {
    return target;
  }
  //
  //
  //
  //
  //
  static userDataAttribs(object) {
    return {};
  }
  static setIndexedAttribute(object, attribName, values, indices) {
  }
  static attribValueIndex(object, index, attribName) {
    return -1;
  }
  //
  //
  //
  //
  //
}
