// @polygonjs/polygonjs/dist/src/core/geometry/modules/cad/CadLoaderSync.js
"use strict";
let _oc;
export class CadLoaderSync {
  static __setOC(oc) {
    _oc = oc;
    this.gp_Ax1 = new oc.gp_Ax1_1();
    this.gp_Ax2 = new oc.gp_Ax2_1();
    this.gp_Dir = new oc.gp_Dir_1();
    this.gp_Pnt2d = new oc.gp_Pnt2d_1();
    this.gp_Vec2d = new oc.gp_Vec2d_1();
    this.gp_Pln = new oc.gp_Pln_1();
    this.gp_Pnt = new oc.gp_Pnt_1();
    this.gp_Quaternion = new oc.gp_Quaternion_1();
    this.gp_Trsf = new oc.gp_Trsf_1();
    this.gp_TrsfT = new oc.gp_Trsf_1();
    this.gp_TrsfR = new oc.gp_Trsf_1();
    this.gp_TrsfS = new oc.gp_Trsf_1();
    this.gp_Vec = new oc.gp_Vec_1();
    this.gp_XYZ = new oc.gp_XYZ_1();
    this.Bnd_Box = new oc.Bnd_Box_1();
    this.TopLoc_Location = new oc.TopLoc_Location_1();
    this.Message_ProgressRange = new oc.Message_ProgressRange_1();
    this.TopTools_ListOfShape = new oc.TopTools_ListOfShape_1();
    this.GProp_GProps = new oc.GProp_GProps_1();
  }
  static oc() {
    return _oc;
  }
  // static TopLoc_Location:TopLoc_Location
}
