// @polygonjs/polygonjs/dist/src/engine/nodes/mat/MeshDepth.js
"use strict";
import { MeshDepthMaterial } from "three";
import { FrontSide } from "three";
import { PrimitiveMatNode } from "./_Base";
import { NodeParamsConfig } from "../utils/params/ParamsConfig";
import {
  AdvancedCommonController,
  AdvancedCommonParamConfig
} from "./utils/AdvancedCommonController";
class MeshDepthMatParamsConfig extends AdvancedCommonParamConfig(NodeParamsConfig) {
}
const ParamsConfig = new MeshDepthMatParamsConfig();
export class MeshDepthMatNode extends PrimitiveMatNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
    this.controllers = {
      advancedCommon: new AdvancedCommonController(this)
    };
    this.controllersList = Object.values(this.controllers);
  }
  static type() {
    return "meshDepth";
  }
  createMaterial() {
    return new MeshDepthMaterial({
      side: FrontSide
    });
  }
  async cook() {
    this._material = this._material || this.createMaterial();
    await Promise.all(this.controllersPromises(this._material));
    this.setMaterial(this._material);
  }
}
