// @polygonjs/polygonjs/dist/src/core/computerVision/hand/Common.js
"use strict";
export var CoreComputerVisionHandAttribute = /* @__PURE__ */ ((CoreComputerVisionHandAttribute2) => {
  CoreComputerVisionHandAttribute2["SELFIE_MODE"] = "CoreComputerVisionHand-selfieMode";
  CoreComputerVisionHandAttribute2["MAX_NUM_HANDS"] = "CoreComputerVisionHand-maxNumHands";
  CoreComputerVisionHandAttribute2["MODEL_COMPLEXITY"] = "CoreComputerVisionHand-modelComplexity";
  CoreComputerVisionHandAttribute2["MIN_DETECTION_CONFIDENCE"] = "CoreComputerVisionHand-minDetectionConfidence";
  CoreComputerVisionHandAttribute2["MAX_TRACKING_CONFIDENCE"] = "CoreComputerVisionHand-minTrackingConfidence";
  CoreComputerVisionHandAttribute2["KEY"] = "CoreComputerVisionHand-key";
  return CoreComputerVisionHandAttribute2;
})(CoreComputerVisionHandAttribute || {});
export var CoreComputerVisionHandIndex = /* @__PURE__ */ ((CoreComputerVisionHandIndex2) => {
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["WRIST"] = 0] = "WRIST";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["THUMB_CMC"] = 1] = "THUMB_CMC";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["THUMB_MCP"] = 2] = "THUMB_MCP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["THUMB_IP"] = 3] = "THUMB_IP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["THUMB_TIP"] = 4] = "THUMB_TIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["INDEX_FINGER_MCP"] = 5] = "INDEX_FINGER_MCP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["INDEX_FINGER_PIP"] = 6] = "INDEX_FINGER_PIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["INDEX_FINGER_DIP"] = 7] = "INDEX_FINGER_DIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["INDEX_FINGER_TIP"] = 8] = "INDEX_FINGER_TIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["MIDDLE_FINGER_MCP"] = 9] = "MIDDLE_FINGER_MCP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["MIDDLE_FINGER_PIP"] = 10] = "MIDDLE_FINGER_PIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["MIDDLE_FINGER_DIP"] = 11] = "MIDDLE_FINGER_DIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["MIDDLE_FINGER_TIP"] = 12] = "MIDDLE_FINGER_TIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["RING_FINGER_MCP"] = 13] = "RING_FINGER_MCP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["RING_FINGER_PIP"] = 14] = "RING_FINGER_PIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["RING_FINGER_DIP"] = 15] = "RING_FINGER_DIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["RING_FINGER_TIP"] = 16] = "RING_FINGER_TIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["PINKY_MCP"] = 17] = "PINKY_MCP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["PINKY_PIP"] = 18] = "PINKY_PIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["PINKY_DIP"] = 19] = "PINKY_DIP";
  CoreComputerVisionHandIndex2[CoreComputerVisionHandIndex2["PINKY_TIP"] = 20] = "PINKY_TIP";
  return CoreComputerVisionHandIndex2;
})(CoreComputerVisionHandIndex || {});
