// @polygonjs/polygonjs/dist/src/core/event/MouseEventType.js
"use strict";
export var MouseEventType = /* @__PURE__ */ ((MouseEventType2) => {
  MouseEventType2["auxclick"] = "auxclick";
  MouseEventType2["click"] = "click";
  MouseEventType2["contextmenu"] = "contextmenu";
  MouseEventType2["dblclick"] = "dblclick";
  MouseEventType2["mousedown"] = "mousedown";
  MouseEventType2["mouseenter"] = "mouseenter";
  MouseEventType2["mouseleave"] = "mouseleave";
  MouseEventType2["mousemove"] = "mousemove";
  MouseEventType2["mouseover"] = "mouseover";
  MouseEventType2["mouseout"] = "mouseout";
  MouseEventType2["mouseup"] = "mouseup";
  MouseEventType2["pointerlockchange"] = "pointerlockchange";
  MouseEventType2["pointerlockerror"] = "pointerlockerror";
  MouseEventType2["select"] = "select";
  MouseEventType2["wheel"] = "wheel";
  return MouseEventType2;
})(MouseEventType || {});
export const ACCEPTED_MOUSE_EVENT_TYPES = [
  "auxclick" /* auxclick */,
  "click" /* click */,
  "contextmenu" /* contextmenu */,
  "dblclick" /* dblclick */,
  "mousedown" /* mousedown */,
  "mouseenter" /* mouseenter */,
  "mouseleave" /* mouseleave */,
  "mousemove" /* mousemove */,
  "mouseover" /* mouseover */,
  "mouseout" /* mouseout */,
  "mouseup" /* mouseup */,
  "pointerlockchange" /* pointerlockchange */,
  "pointerlockerror" /* pointerlockerror */,
  "select" /* select */,
  "wheel" /* wheel */
];
