// @polygonjs/polygonjs/dist/src/engine/nodes/js/ObjectDispatchEvent.js
"use strict";
import { TRIGGER_CONNECTION_NAME, TypedJsNode } from "./_Base";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { JsConnectionPoint, JsConnectionPointType, JS_CONNECTION_POINT_IN_NODE_DEF } from "../utils/io/connections/Js";
import { inputObject3D } from "./_BaseObject3D";
import { Poly } from "../../Poly";
const CONNECTION_OPTIONS = JS_CONNECTION_POINT_IN_NODE_DEF;
class ObjectDispatchEventJsParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    /** @param event name */
    this.eventName = ParamConfig.STRING("my-event");
  }
}
const ParamsConfig = new ObjectDispatchEventJsParamsConfig();
export class ObjectDispatchEventJsNode extends TypedJsNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return "objectDispatchEvent";
  }
  initializeNode() {
    super.initializeNode();
    this.io.inputs.setNamedInputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER, CONNECTION_OPTIONS),
      new JsConnectionPoint(JsConnectionPointType.OBJECT_3D, JsConnectionPointType.OBJECT_3D, CONNECTION_OPTIONS)
    ]);
    this.io.outputs.setNamedOutputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER),
      new JsConnectionPoint(JsConnectionPointType.OBJECT_3D, JsConnectionPointType.OBJECT_3D, CONNECTION_OPTIONS)
    ]);
  }
  setTriggerableLines(shadersCollectionController) {
    const object3D = inputObject3D(this, shadersCollectionController);
    const eventName = this.variableForInputParam(shadersCollectionController, this.p.eventName);
    const func = Poly.namedFunctionsRegister.getFunction("objectDispatchEvent", this, shadersCollectionController);
    const bodyLine = func.asString(object3D, eventName);
    shadersCollectionController.addTriggerableLines(this, [bodyLine]);
  }
}
