// @polygonjs/polygonjs/dist/src/engine/nodes/js/Object3DUpdateWorldMatrix.js
"use strict";
import { TRIGGER_CONNECTION_NAME, TypedJsNode } from "./_Base";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { JsConnectionPoint, JsConnectionPointType, JS_CONNECTION_POINT_IN_NODE_DEF } from "../utils/io/connections/Js";
import { inputObject3D } from "./_BaseObject3D";
import { Poly } from "../../Poly";
const CONNECTION_OPTIONS = JS_CONNECTION_POINT_IN_NODE_DEF;
class Object3DUpdateWorldMatrixJsParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    /** @param updates the matrix of the parents */
    this.updateParents = ParamConfig.BOOLEAN(1);
    /** @param updates the matrix of the children */
    this.updateChildren = ParamConfig.BOOLEAN(1);
  }
}
const ParamsConfig = new Object3DUpdateWorldMatrixJsParamsConfig();
export class Object3DUpdateWorldMatrixJsNode extends TypedJsNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return "object3DUpdateWorldMatrix";
  }
  initializeNode() {
    super.initializeNode();
    this.io.inputs.setNamedInputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER, CONNECTION_OPTIONS),
      new JsConnectionPoint(JsConnectionPointType.OBJECT_3D, JsConnectionPointType.OBJECT_3D, CONNECTION_OPTIONS)
    ]);
    this.io.outputs.setNamedOutputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER)
    ]);
  }
  setTriggerableLines(shadersCollectionController) {
    const object3D = inputObject3D(this, shadersCollectionController);
    const updateParents = this.variableForInputParam(shadersCollectionController, this.p.updateParents);
    const updateChildren = this.variableForInputParam(shadersCollectionController, this.p.updateChildren);
    const func = Poly.namedFunctionsRegister.getFunction(
      "objectUpdateWorldMatrix",
      this,
      shadersCollectionController
    );
    const bodyLine = func.asString(object3D, updateParents, updateChildren);
    shadersCollectionController.addTriggerableLines(this, [bodyLine]);
  }
}
