// @polygonjs/polygonjs/dist/src/engine/nodes/js/OnMapboxCameraMoveEnd.js
"use strict";
import { TRIGGER_CONNECTION_NAME, TypedJsNode } from "./_Base";
import { NodeParamsConfig } from "../utils/params/ParamsConfig";
import { JsConnectionPoint, JsConnectionPointType } from "../utils/io/connections/Js";
import { JsType } from "../../poly/registers/nodes/types/Js";
class OnMapboxCameraMoveEndJsParamsConfig extends NodeParamsConfig {
}
const ParamsConfig = new OnMapboxCameraMoveEndJsParamsConfig();
export class OnMapboxCameraMoveEndJsNode extends TypedJsNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return JsType.ON_MAPBOX_CAMERA_MOVE_END;
  }
  isTriggering() {
    return true;
  }
  initializeNode() {
    this.io.outputs.setNamedOutputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER)
    ]);
  }
  setTriggeringLines(shadersCollectionController, triggeredMethods) {
    shadersCollectionController.addTriggeringLines(this, [triggeredMethods], {
      gatherable: true
    });
  }
}
