// @polygonjs/polygonjs/dist/src/engine/nodes/js/SetPointPosition.js
"use strict";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { JsType } from "../../poly/registers/nodes/types/Js";
import { Poly } from "../../Poly";
import {
  inputObject3D,
  inputPointIndex,
  vector3OutputFromParam,
  floatOutputFromParam,
  integerOutputFromParam
} from "./_BaseObject3D";
import { BaseTriggerAndObjectJsNode } from "./_BaseTriggerAndObject";
import { JS_CONNECTION_POINT_IN_NODE_DEF, JsConnectionPoint, JsConnectionPointType } from "../utils/io/connections/Js";
const CONNECTION_OPTIONS = JS_CONNECTION_POINT_IN_NODE_DEF;
class SetPointPositionJsParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    /** @param point index */
    this.ptnum = ParamConfig.INTEGER(0);
    /** @param target position */
    this.position = ParamConfig.VECTOR3([0, 0, 0]);
    /** @param lerp factor */
    this.lerp = ParamConfig.FLOAT(1);
  }
}
const ParamsConfig = new SetPointPositionJsParamsConfig();
export class SetPointPositionJsNode extends BaseTriggerAndObjectJsNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return JsType.SET_POINT_POSITION;
  }
  _additionalOutputs() {
    return [
      new JsConnectionPoint("ptnum", JsConnectionPointType.INT, CONNECTION_OPTIONS),
      new JsConnectionPoint("position", JsConnectionPointType.VECTOR3, CONNECTION_OPTIONS),
      new JsConnectionPoint("lerp", JsConnectionPointType.FLOAT, CONNECTION_OPTIONS)
    ];
  }
  setLines(linesController) {
    super.setLines(linesController);
    integerOutputFromParam(this, this.p.ptnum, linesController);
    vector3OutputFromParam(this, this.p.position, linesController);
    floatOutputFromParam(this, this.p.lerp, linesController);
  }
  setTriggerableLines(linesController) {
    const object3D = inputObject3D(this, linesController);
    const ptnum = inputPointIndex(this, linesController);
    const position = this.variableForInputParam(linesController, this.p.position);
    const lerp = this.variableForInputParam(linesController, this.p.lerp);
    const func = Poly.namedFunctionsRegister.getFunction("setPointPosition", this, linesController);
    const bodyLine = func.asString(object3D, ptnum, position, lerp);
    linesController.addTriggerableLines(this, [bodyLine]);
  }
}
