// @polygonjs/polygonjs/dist/src/engine/nodes/js/SetPhysicsRBDCapsuleProperty.js
"use strict";
import { ParamConfig } from "../utils/params/ParamsConfig";
import { TRIGGER_CONNECTION_NAME, TypedJsNode } from "./_Base";
import { NodeParamsConfig } from "../utils/params/ParamsConfig";
import { JsConnectionPoint, JsConnectionPointType, JS_CONNECTION_POINT_IN_NODE_DEF } from "../utils/io/connections/Js";
import { inputObject3D, setObject3DOutputLine } from "./_BaseObject3D";
import { Poly } from "../../Poly";
const CONNECTION_OPTIONS = JS_CONNECTION_POINT_IN_NODE_DEF;
class SetPhysicsRBDCapsulePropertyJsParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    /** @param target scale */
    this.scale = ParamConfig.FLOAT(1);
    /** @param lerp factor */
    this.lerp = ParamConfig.FLOAT(1);
    /** @param sets if the matrix should be updated as the animation progresses */
    this.updateMatrix = ParamConfig.BOOLEAN(1);
  }
}
const ParamsConfig = new SetPhysicsRBDCapsulePropertyJsParamsConfig();
export class SetPhysicsRBDCapsulePropertyJsNode extends TypedJsNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return "setPhysicsRBDCapsuleProperty";
  }
  initializeNode() {
    this.io.inputs.setNamedInputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER, CONNECTION_OPTIONS),
      new JsConnectionPoint(JsConnectionPointType.OBJECT_3D, JsConnectionPointType.OBJECT_3D, CONNECTION_OPTIONS)
    ]);
    this.io.outputs.setNamedOutputConnectionPoints([
      new JsConnectionPoint(TRIGGER_CONNECTION_NAME, JsConnectionPointType.TRIGGER),
      new JsConnectionPoint(JsConnectionPointType.OBJECT_3D, JsConnectionPointType.OBJECT_3D, CONNECTION_OPTIONS)
    ]);
  }
  setLines(linesController) {
    setObject3DOutputLine(this, linesController);
  }
  setTriggerableLines(shadersCollectionController) {
    const object3D = inputObject3D(this, shadersCollectionController);
    const scale = this.variableForInputParam(shadersCollectionController, this.p.scale);
    const lerp = this.variableForInputParam(shadersCollectionController, this.p.lerp);
    const updateMatrix = this.variableForInputParam(shadersCollectionController, this.p.updateMatrix);
    const func = Poly.namedFunctionsRegister.getFunction(
      "setPhysicsRBDCapsuleProperty",
      this,
      shadersCollectionController
    );
    const bodyLine = func.asString(object3D, scale, lerp, updateMatrix);
    shadersCollectionController.addTriggerableLines(this, [bodyLine]);
  }
}
