// @polygonjs/polygonjs/dist/src/engine/functions/_SetObjectProperty.js
"use strict";
import { touchObjectProperty, GetObjectPropertyJsNodeInputName } from "../../core/reactivity/ObjectPropertyReactivity";
import { ObjectNamedFunction0, ObjectNamedFunction1, ObjectNamedFunction2 } from "./_Base";
export class setObjectCastShadow extends ObjectNamedFunction1 {
  static type() {
    return "setObjectCastShadow";
  }
  func(object3D, castShadow) {
    object3D.castShadow = castShadow;
    touchObjectProperty(object3D, GetObjectPropertyJsNodeInputName.castShadow);
  }
}
export class setObjectReceiveShadow extends ObjectNamedFunction1 {
  static type() {
    return "setObjectReceiveShadow";
  }
  func(object3D, receiveShadow) {
    object3D.receiveShadow = receiveShadow;
    touchObjectProperty(object3D, GetObjectPropertyJsNodeInputName.receiveShadow);
  }
}
export class setObjectFrustumCulled extends ObjectNamedFunction1 {
  static type() {
    return "setObjectFrustumCulled";
  }
  func(object3D, frustumCulled) {
    object3D.frustumCulled = frustumCulled;
    touchObjectProperty(object3D, "frustumCulled");
  }
}
export class setObjectMatrix extends ObjectNamedFunction1 {
  static type() {
    return "setObjectMatrix";
  }
  func(object3D, matrix) {
    object3D.matrix.copy(matrix);
    object3D.matrix.decompose(object3D.position, object3D.quaternion, object3D.scale);
    touchObjectProperty(object3D, GetObjectPropertyJsNodeInputName.matrix);
  }
}
export class setObjectMatrixAutoUpdate extends ObjectNamedFunction1 {
  static type() {
    return "setObjectMatrixAutoUpdate";
  }
  func(object3D, matrixAutoUpdate) {
    object3D.matrixAutoUpdate = matrixAutoUpdate;
    touchObjectProperty(object3D, GetObjectPropertyJsNodeInputName.matrixAutoUpdate);
  }
}
export class setObjectVisible extends ObjectNamedFunction1 {
  static type() {
    return "setObjectVisible";
  }
  func(object3D, visible) {
    object3D.visible = visible;
    touchObjectProperty(object3D, GetObjectPropertyJsNodeInputName.visible);
  }
}
export class objectUpdateMatrix extends ObjectNamedFunction0 {
  static type() {
    return "objectUpdateMatrix";
  }
  func(object3D) {
    object3D.updateMatrix();
    touchObjectProperty(object3D, GetObjectPropertyJsNodeInputName.matrix);
  }
}
export class objectUpdateWorldMatrix extends ObjectNamedFunction2 {
  static type() {
    return "objectUpdateWorldMatrix";
  }
  func(object3D, updateParents, updateChildren) {
    object3D.updateWorldMatrix(updateParents, updateChildren);
    touchObjectProperty(object3D, GetObjectPropertyJsNodeInputName.matrix);
  }
}
