// @polygonjs/polygonjs/dist/src/engine/nodes/js/code/assemblers/objectBuilder/ObjectVariables.js
"use strict";
export var ObjectVariableLight = /* @__PURE__ */ ((ObjectVariableLight2) => {
  ObjectVariableLight2["INTENSITY"] = "intensity";
  ObjectVariableLight2["COLOR"] = "color";
  return ObjectVariableLight2;
})(ObjectVariableLight || {});
export var ObjectVariableAreaLight = /* @__PURE__ */ ((ObjectVariableAreaLight2) => {
  ObjectVariableAreaLight2["WIDTH"] = "width";
  ObjectVariableAreaLight2["HEIGHT"] = "height";
  return ObjectVariableAreaLight2;
})(ObjectVariableAreaLight || {});
export var ObjectVariableDirectionalLight = /* @__PURE__ */ ((ObjectVariableDirectionalLight2) => {
  ObjectVariableDirectionalLight2["DISTANCE"] = "distance";
  ObjectVariableDirectionalLight2["SHADOW_BIAS"] = "shadowBias";
  ObjectVariableDirectionalLight2["SHADOW_RADIUS"] = "shadowRadius";
  return ObjectVariableDirectionalLight2;
})(ObjectVariableDirectionalLight || {});
export var ObjectVariableHemisphereLight = /* @__PURE__ */ ((ObjectVariableHemisphereLight2) => {
  ObjectVariableHemisphereLight2["SKY_COLOR"] = "skyColor";
  ObjectVariableHemisphereLight2["GROUND_COLOR"] = "groundColor";
  return ObjectVariableHemisphereLight2;
})(ObjectVariableHemisphereLight || {});
export var ObjectVariablePointLight = /* @__PURE__ */ ((ObjectVariablePointLight2) => {
  ObjectVariablePointLight2["DECAY"] = "decay";
  ObjectVariablePointLight2["DISTANCE"] = "distance";
  ObjectVariablePointLight2["SHADOW_BIAS"] = "shadowBias";
  ObjectVariablePointLight2["SHADOW_NEAR"] = "shadowNear";
  ObjectVariablePointLight2["SHADOW_FAR"] = "shadowFar";
  return ObjectVariablePointLight2;
})(ObjectVariablePointLight || {});
export var ObjectVariableSpotLight = /* @__PURE__ */ ((ObjectVariableSpotLight2) => {
  ObjectVariableSpotLight2["ANGLE"] = "angle";
  ObjectVariableSpotLight2["PENUMBRA"] = "penumbra";
  ObjectVariableSpotLight2["DECAY"] = "decay";
  ObjectVariableSpotLight2["DISTANCE"] = "distance";
  ObjectVariableSpotLight2["SHADOW_BIAS"] = "shadowBias";
  ObjectVariableSpotLight2["SHADOW_NEAR"] = "shadowNear";
  ObjectVariableSpotLight2["SHADOW_FAR"] = "shadowFar";
  ObjectVariableSpotLight2["SHADOW_RADIUS"] = "shadowRadius";
  return ObjectVariableSpotLight2;
})(ObjectVariableSpotLight || {});
