// @polygonjs/polygonjs/dist/src/engine/nodes/mat/MeshBasic.js
"use strict";
import { MeshBasicMaterial } from "three";
import { FrontSide } from "three";
import { PrimitiveMatNode } from "./_Base";
import { NodeParamsConfig } from "../utils/params/ParamsConfig";
import { ColorsController, ColorParamConfig } from "./utils/ColorsController";
import { FogController, FogParamConfig } from "./utils/FogController";
import {
  AdvancedCommonController,
  AdvancedCommonParamConfig
} from "./utils/AdvancedCommonController";
import { TextureMapController, MapParamConfig } from "./utils/TextureMapController";
import {
  TextureAlphaMapController,
  AlphaMapParamConfig
} from "./utils/TextureAlphaMapController";
import { TextureAOMapController, AOMapParamConfig } from "./utils/TextureAOMapController";
import {
  TextureEnvMapSimpleController,
  EnvMapSimpleParamConfig
} from "./utils/TextureEnvMapSimpleController";
import {
  TextureLightMapController,
  LightMapParamConfig
} from "./utils/TextureLightMapController";
import { WireframeController, WireframeParamConfig } from "./utils/WireframeController";
import { DefaultFolderParamConfig } from "./utils/DefaultFolder";
import { TexturesFolderParamConfig } from "./utils/TexturesFolder";
import { AdvancedFolderParamConfig } from "./utils/AdvancedFolder";
import { MatType } from "../../poly/registers/nodes/types/Mat";
class MeshBasicMatParamsConfig extends FogParamConfig(
  WireframeParamConfig(
    AdvancedCommonParamConfig(
      /* advanced */
      AdvancedFolderParamConfig(
        LightMapParamConfig(
          EnvMapSimpleParamConfig(
            AOMapParamConfig(
              AlphaMapParamConfig(
                MapParamConfig(
                  /* textures */
                  TexturesFolderParamConfig(
                    ColorParamConfig(DefaultFolderParamConfig(NodeParamsConfig))
                  )
                )
              )
            )
          )
        )
      )
    )
  )
) {
}
const ParamsConfig = new MeshBasicMatParamsConfig();
export class MeshBasicMatNode extends PrimitiveMatNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
    this.controllers = {
      colors: new ColorsController(this),
      advancedCommon: new AdvancedCommonController(this),
      alphaMap: new TextureAlphaMapController(this),
      aoMap: new TextureAOMapController(this),
      envMap: new TextureEnvMapSimpleController(this),
      fog: new FogController(this),
      lightMap: new TextureLightMapController(this),
      map: new TextureMapController(this),
      wireframe: new WireframeController(this)
    };
    this.controllersList = Object.values(this.controllers);
  }
  static type() {
    return MatType.MESH_BASIC;
  }
  createMaterial() {
    return new MeshBasicMaterial({
      vertexColors: false,
      side: FrontSide,
      color: 16777215,
      opacity: 1
    });
  }
  async cook() {
    this._material = this._material || this.createMaterial();
    await Promise.all(this.controllersPromises(this._material));
    this.setMaterial(this._material);
  }
}
