// @polygonjs/polygonjs/dist/src/engine/nodes/mat/MeshPhong.js
"use strict";
import { MeshPhongMaterial } from "three";
import { FrontSide } from "three";
import { PrimitiveMatNode } from "./_Base";
import { NodeParamsConfig } from "../utils/params/ParamsConfig";
import { ColorsController, ColorParamConfig } from "./utils/ColorsController";
import {
  AdvancedCommonController,
  AdvancedCommonParamConfig
} from "./utils/AdvancedCommonController";
import { TextureMapController, MapParamConfig } from "./utils/TextureMapController";
import {
  TextureAlphaMapController,
  AlphaMapParamConfig
} from "./utils/TextureAlphaMapController";
import {
  TextureBumpMapController,
  BumpMapParamConfig
} from "./utils/TextureBumpMapController";
import {
  TextureNormalMapController,
  NormalMapParamConfig
} from "./utils/TextureNormalMapController";
import {
  TextureSpecularMapController,
  SpecularMapParamConfig
} from "./utils/TextureSpecularMapController";
import {
  TextureEnvMapSimpleController,
  EnvMapSimpleParamConfig
} from "./utils/TextureEnvMapSimpleController";
import {
  TextureEmissiveMapController,
  EmissiveMapParamConfig
} from "./utils/TextureEmissiveMapController";
import {
  TextureDisplacementMapController,
  DisplacementMapParamConfig
} from "./utils/TextureDisplacementMapController";
import {
  TextureLightMapController,
  LightMapParamConfig
} from "./utils/TextureLightMapController";
import { TextureAOMapController, AOMapParamConfig } from "./utils/TextureAOMapController";
import { FlatShadingController, FlatShadingParamConfig } from "./utils/FlatShadingController";
import { WireframeController, WireframeParamConfig } from "./utils/WireframeController";
import { MatType } from "../../poly/registers/nodes/types/Mat";
import { FogController, FogParamConfig } from "./utils/FogController";
import { DefaultFolderParamConfig } from "./utils/DefaultFolder";
import { TexturesFolderParamConfig } from "./utils/TexturesFolder";
import { AdvancedFolderParamConfig } from "./utils/AdvancedFolder";
class MeshPhongMatParamsConfig extends FogParamConfig(
  WireframeParamConfig(
    AdvancedCommonParamConfig(
      FlatShadingParamConfig(
        /* advanced */
        AdvancedFolderParamConfig(
          SpecularMapParamConfig(
            NormalMapParamConfig(
              LightMapParamConfig(
                EnvMapSimpleParamConfig(
                  EmissiveMapParamConfig(
                    DisplacementMapParamConfig(
                      BumpMapParamConfig(
                        AOMapParamConfig(
                          AlphaMapParamConfig(
                            MapParamConfig(
                              /* textures */
                              TexturesFolderParamConfig(
                                ColorParamConfig(
                                  DefaultFolderParamConfig(NodeParamsConfig)
                                )
                              )
                            )
                          )
                        )
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    )
  )
) {
}
const ParamsConfig = new MeshPhongMatParamsConfig();
export class MeshPhongMatNode extends PrimitiveMatNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
    this.controllers = {
      colors: new ColorsController(this),
      advancedCommon: new AdvancedCommonController(this),
      alphaMap: new TextureAlphaMapController(this),
      aoMap: new TextureAOMapController(this),
      bumpMap: new TextureBumpMapController(this),
      displacementMap: new TextureDisplacementMapController(this),
      emissiveMap: new TextureEmissiveMapController(this),
      envMap: new TextureEnvMapSimpleController(this),
      fog: new FogController(this),
      flatShading: new FlatShadingController(this),
      lightMap: new TextureLightMapController(this),
      map: new TextureMapController(this),
      normalMap: new TextureNormalMapController(this),
      specularMap: new TextureSpecularMapController(this),
      wireframe: new WireframeController(this)
    };
    this.controllersList = Object.values(this.controllers);
  }
  static type() {
    return MatType.MESH_PHONG;
  }
  createMaterial() {
    return new MeshPhongMaterial({
      vertexColors: false,
      side: FrontSide,
      color: 16777215,
      opacity: 1
    });
  }
  async cook() {
    this._material = this._material || this.createMaterial();
    await Promise.all(this.controllersPromises(this._material));
    this.setMaterial(this._material);
  }
}
