// @polygonjs/polygonjs/dist/src/engine/nodes/mat/MeshLambert.js
"use strict";
import { MeshLambertMaterial } from "three";
import { FrontSide } from "three";
import { PrimitiveMatNode } from "./_Base";
import { NodeParamsConfig } from "../utils/params/ParamsConfig";
import { ColorsController, ColorParamConfig } from "./utils/ColorsController";
import {
  AdvancedCommonController,
  AdvancedCommonParamConfig
} from "./utils/AdvancedCommonController";
import { TextureMapController, MapParamConfig } from "./utils/TextureMapController";
import {
  TextureAlphaMapController,
  AlphaMapParamConfig
} from "./utils/TextureAlphaMapController";
import {
  TextureEnvMapSimpleController,
  EnvMapSimpleParamConfig
} from "./utils/TextureEnvMapSimpleController";
import {
  TextureLightMapController,
  LightMapParamConfig
} from "./utils/TextureLightMapController";
import {
  TextureEmissiveMapController,
  EmissiveMapParamConfig
} from "./utils/TextureEmissiveMapController";
import { FlatShadingController, FlatShadingParamConfig } from "./utils/FlatShadingController";
import { TextureAOMapController, AOMapParamConfig } from "./utils/TextureAOMapController";
import { WireframeController, WireframeParamConfig } from "./utils/WireframeController";
import { FogController, FogParamConfig } from "./utils/FogController";
import { DefaultFolderParamConfig } from "./utils/DefaultFolder";
import { TexturesFolderParamConfig } from "./utils/TexturesFolder";
import { AdvancedFolderParamConfig } from "./utils/AdvancedFolder";
import { MatType } from "../../poly/registers/nodes/types/Mat";
class MeshLambertMatParamsConfig extends FogParamConfig(
  WireframeParamConfig(
    AdvancedCommonParamConfig(
      FlatShadingParamConfig(
        /* advanced */
        AdvancedFolderParamConfig(
          LightMapParamConfig(
            EnvMapSimpleParamConfig(
              EmissiveMapParamConfig(
                AOMapParamConfig(
                  AlphaMapParamConfig(
                    MapParamConfig(
                      /* textures */
                      TexturesFolderParamConfig(
                        ColorParamConfig(DefaultFolderParamConfig(NodeParamsConfig))
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    )
  )
) {
}
const ParamsConfig = new MeshLambertMatParamsConfig();
export class MeshLambertMatNode extends PrimitiveMatNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
    this.controllers = {
      colors: new ColorsController(this),
      advancedCommon: new AdvancedCommonController(this),
      alphaMap: new TextureAlphaMapController(this),
      aoMap: new TextureAOMapController(this),
      emissiveMap: new TextureEmissiveMapController(this),
      envMap: new TextureEnvMapSimpleController(this),
      fog: new FogController(this),
      flatShading: new FlatShadingController(this),
      lightMap: new TextureLightMapController(this),
      map: new TextureMapController(this),
      wireframe: new WireframeController(this)
    };
    this.controllersList = Object.values(this.controllers);
  }
  static type() {
    return MatType.MESH_LAMBERT;
  }
  createMaterial() {
    return new MeshLambertMaterial({
      vertexColors: false,
      side: FrontSide,
      color: 16777215,
      opacity: 1
    });
  }
  async cook() {
    this._material = this._material || this.createMaterial();
    await Promise.all(this.controllersPromises(this._material));
    this.setMaterial(this._material);
  }
}
