// @polygonjs/polygonjs/dist/src/engine/nodes/event/Drag.js
"use strict";
import { EventConnectionPoint, EventConnectionPointType } from "../utils/io/connections/Event";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { TypedInputEventNode, EVENT_PARAM_OPTIONS } from "./_BaseInput";
import { isBooleanTrue } from "../../../core/BooleanValue";
import { EventInputType } from "../../poly/registers/nodes/types/Event";
import { CoreEventEmitter, EVENT_EMITTERS, EVENT_EMITTER_PARAM_MENU_OPTIONS } from "../../../core/event/CoreEventEmitter";
import { ACCEPTED_DRAG_EVENT_TYPES } from "../../../core/event/DragEventType";
class DragEventParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    /** @param toggle on to allow any event to be listened to */
    this.active = ParamConfig.BOOLEAN(true, {
      callback: (node) => {
        DragEventNode.PARAM_CALLBACK_updateRegister(node);
      },
      separatorAfter: true
    });
    /** @param set which element triggers the event */
    this.element = ParamConfig.INTEGER(EVENT_EMITTERS.indexOf(CoreEventEmitter.CANVAS), {
      ...EVENT_EMITTER_PARAM_MENU_OPTIONS,
      separatorAfter: true
    });
    /** @param toggle on to listen to dragover events */
    this.dragover = ParamConfig.BOOLEAN(1, EVENT_PARAM_OPTIONS);
    /** @param requires ctrlKey */
    this.ctrlKey = ParamConfig.BOOLEAN(0, { ...EVENT_PARAM_OPTIONS, separatorBefore: true });
    /** @param requires altKey */
    this.altKey = ParamConfig.BOOLEAN(0, EVENT_PARAM_OPTIONS);
    /** @param requires shiftKey */
    this.shiftKey = ParamConfig.BOOLEAN(0, EVENT_PARAM_OPTIONS);
    /** @param requires metaKey */
    this.metaKey = ParamConfig.BOOLEAN(0, EVENT_PARAM_OPTIONS);
  }
}
const ParamsConfig = new DragEventParamsConfig();
export class DragEventNode extends TypedInputEventNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return EventInputType.DRAG;
  }
  acceptedEventTypes() {
    return /* @__PURE__ */ new Set([...ACCEPTED_DRAG_EVENT_TYPES]);
  }
  initializeNode() {
    this.io.outputs.setNamedOutputConnectionPoints(
      ACCEPTED_DRAG_EVENT_TYPES.map((event_type) => {
        return new EventConnectionPoint(event_type, EventConnectionPointType.DRAG);
      })
    );
  }
  processEvent(eventContext) {
    if (!this.pv.active) {
      return;
    }
    const event = eventContext.event;
    if (!event) {
      return;
    }
    if (event.ctrlKey != isBooleanTrue(this.pv.ctrlKey)) {
      return;
    }
    if (event.shiftKey != isBooleanTrue(this.pv.shiftKey)) {
      return;
    }
    if (event.altKey != isBooleanTrue(this.pv.altKey)) {
      return;
    }
    if (event.metaKey != isBooleanTrue(this.pv.metaKey)) {
      return;
    }
    this.dispatchEventToOutput(event.type, eventContext);
  }
}
