// @polygonjs/polygonjs/dist/src/engine/nodes/event/Touch.js
"use strict";
import { EventConnectionPoint, EventConnectionPointType } from "../utils/io/connections/Event";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { TypedInputEventNode, EVENT_PARAM_OPTIONS } from "./_BaseInput";
import { EventInputType } from "../../poly/registers/nodes/types/Event";
import { CoreEventEmitter, EVENT_EMITTERS, EVENT_EMITTER_PARAM_MENU_OPTIONS } from "../../../core/event/CoreEventEmitter";
import { ACCEPTED_TOUCH_EVENT_TYPES } from "../../../core/event/TouchEventType";
class TouchEventParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    /** @param toggle on to allow any event to be listened to */
    this.active = ParamConfig.BOOLEAN(true, {
      callback: (node) => {
        TouchEventNode.PARAM_CALLBACK_updateRegister(node);
      },
      separatorAfter: true
    });
    /** @param set which element triggers the event */
    this.element = ParamConfig.INTEGER(EVENT_EMITTERS.indexOf(CoreEventEmitter.CANVAS), {
      ...EVENT_EMITTER_PARAM_MENU_OPTIONS,
      separatorAfter: true
    });
    /** @param toggle on to listen to touchstart events */
    this.touchstart = ParamConfig.BOOLEAN(1, EVENT_PARAM_OPTIONS);
    /** @param toggle on to listen to touchmove events */
    this.touchmove = ParamConfig.BOOLEAN(0, EVENT_PARAM_OPTIONS);
    /** @param toggle on to listen to touchend events */
    this.touchend = ParamConfig.BOOLEAN(0, EVENT_PARAM_OPTIONS);
  }
}
const ParamsConfig = new TouchEventParamsConfig();
export class TouchEventNode extends TypedInputEventNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return EventInputType.TOUCH;
  }
  acceptedEventTypes() {
    return /* @__PURE__ */ new Set([...ACCEPTED_TOUCH_EVENT_TYPES]);
  }
  initializeNode() {
    this.io.outputs.setNamedOutputConnectionPoints(
      ACCEPTED_TOUCH_EVENT_TYPES.map((event_type) => {
        return new EventConnectionPoint(event_type, EventConnectionPointType.DRAG);
      })
    );
  }
  processEvent(eventContext) {
    if (!this.pv.active) {
      return;
    }
    if (!eventContext.event) {
      return;
    }
    const event = eventContext.event;
    if (!event) {
      return;
    }
    this.dispatchEventToOutput(event.type, eventContext);
  }
}
