// @polygonjs/polygonjs/dist/src/core/thirdParty/Mapbox/MapboxCameraAttribute.js
"use strict";
export var MapboxCameraAttribute = /* @__PURE__ */ ((MapboxCameraAttribute2) => {
  MapboxCameraAttribute2["STYLE"] = "mapbox-camera-attrib-style";
  MapboxCameraAttribute2["LONGITUDE"] = "mapbox-camera-attrib-longitude";
  MapboxCameraAttribute2["LATITUDE"] = "mapbox-camera-attrib-latitude";
  MapboxCameraAttribute2["ZOOM"] = "mapbox-camera-attrib-zoom";
  MapboxCameraAttribute2["MIN_ZOOM"] = "mapbox-camera-attrib-minZoom";
  MapboxCameraAttribute2["MAX_ZOOM"] = "mapbox-camera-attrib-maxZoom";
  MapboxCameraAttribute2["PITCH"] = "mapbox-camera-attrib-pitch";
  MapboxCameraAttribute2["BEARING"] = "mapbox-camera-attrib-bearing";
  MapboxCameraAttribute2["ALLOW_DRAG_ROTATE"] = "mapbox-camera-attrib-allowDragRotate";
  MapboxCameraAttribute2["ADD_ZOOM_CONTROL"] = "mapbox-camera-attrib-addZoomControl";
  MapboxCameraAttribute2["LAYER_BUILDINGS"] = "mapbox-camera-attrib-tlayerBuildings";
  MapboxCameraAttribute2["LAYER_3D"] = "mapbox-camera-attrib-tlayer3D";
  MapboxCameraAttribute2["LAYER_SKY"] = "mapbox-camera-attrib-tlayerSky";
  return MapboxCameraAttribute2;
})(MapboxCameraAttribute || {});
