// @polygonjs/polygonjs/dist/src/engine/nodes/js/code/assemblers/entityBuilder/EntityBuilderAssemblerCommon.js
"use strict";
export var EntityBuilderAssemblerConstant = /* @__PURE__ */ ((EntityBuilderAssemblerConstant2) => {
  EntityBuilderAssemblerConstant2["OBJECT"] = "entityContainer.object";
  EntityBuilderAssemblerConstant2["ENTITY_CONTAINER"] = "entityContainer";
  EntityBuilderAssemblerConstant2["POSITION"] = "entityContainer.position";
  EntityBuilderAssemblerConstant2["NORMAL"] = "entityContainer.normal";
  EntityBuilderAssemblerConstant2["INDEX"] = "entityContainer.index";
  EntityBuilderAssemblerConstant2["OBJNUM"] = "entityContainer.objnum";
  EntityBuilderAssemblerConstant2["NORMALS_UPDATED"] = "entityContainer.normalsUpdated";
  EntityBuilderAssemblerConstant2["PRIMITIVE_GRAPH"] = "entityContainer.primitiveGraph";
  EntityBuilderAssemblerConstant2["ATTRIBUTES_DICT"] = "attributesDict";
  EntityBuilderAssemblerConstant2["MATERIAL"] = "null";
  return EntityBuilderAssemblerConstant2;
})(EntityBuilderAssemblerConstant || {});
export var EntityVariable = /* @__PURE__ */ ((EntityVariable2) => {
  EntityVariable2["POSITION"] = "position";
  EntityVariable2["NORMAL"] = "normal";
  EntityVariable2["INDEX"] = "index";
  EntityVariable2["OBJNUM"] = "objnum";
  return EntityVariable2;
})(EntityVariable || {});
