// @polygonjs/polygonjs/dist/src/core/render/Common.js
"use strict";
export var PowerPreference = /* @__PURE__ */ ((PowerPreference2) => {
  PowerPreference2["HIGH"] = "high-performance";
  PowerPreference2["LOW"] = "low-power";
  PowerPreference2["DEFAULT"] = "default";
  return PowerPreference2;
})(PowerPreference || {});
export const POWER_PREFERENCES = [
  "high-performance" /* HIGH */,
  "low-power" /* LOW */,
  "default" /* DEFAULT */
];
export var RendererPrecision = /* @__PURE__ */ ((RendererPrecision2) => {
  RendererPrecision2["HIGH"] = "highp";
  RendererPrecision2["MEDIUM"] = "mediump";
  RendererPrecision2["LOW"] = "lowp";
  return RendererPrecision2;
})(RendererPrecision || {});
export const RENDERER_PRECISIONS = [
  "highp" /* HIGH */,
  "mediump" /* MEDIUM */,
  "lowp" /* LOW */
];
export const WEBGL_RENDERER_DEFAULT_PARAMS = {
  alpha: true,
  precision: "highp" /* HIGH */,
  premultipliedAlpha: true,
  antialias: true,
  preserveDrawingBuffer: false,
  powerPreference: "default" /* DEFAULT */,
  depth: true,
  stencil: true,
  logarithmicDepthBuffer: false
};
