// @polygonjs/polygonjs/dist/src/engine/nodes/sop/CADTriangulate.js
"use strict";
import { CADSopNode } from "./_BaseCAD";
import { NodeParamsConfig } from "../utils/params/ParamsConfig";
import { CoreType } from "../../../core/Type";
import { SopType } from "../../poly/registers/nodes/types/Sop";
import { SOPCADTesselationParamConfig } from "../../../core/geometry/modules/cad/utils/TesselationParamsConfig";
class CADTriangulateSopParamsConfig extends SOPCADTesselationParamConfig(NodeParamsConfig) {
}
const ParamsConfig = new CADTriangulateSopParamsConfig();
export class CADTriangulateSopNode extends CADSopNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return SopType.CAD_TRIANGULATE;
  }
  initializeNode() {
    this.io.inputs.setCount(1);
  }
  async cook(inputCoreGroups) {
    const cadObjects = inputCoreGroups[0].cadObjects();
    if (cadObjects) {
      const newObjects = [];
      for (const cadObject of cadObjects) {
        const objects = cadObject.toObject3D(this.pv, this);
        if (objects) {
          if (CoreType.isArray(objects)) {
            newObjects.push(...objects);
          } else {
            newObjects.push(objects);
          }
        }
      }
      this.setObjects(newObjects);
    } else {
      this.setObjects([]);
    }
  }
}
