// @polygonjs/polygonjs/dist/src/engine/nodes/sop/ExporterUSDZ.js
"use strict";
import { ExporterSopNode, BaseExporterSopParamsConfig } from "./_BaseExporter";
import { USDZExporter } from "three/examples/jsm/exporters/USDZExporter";
import { SopExporter } from "../../poly/registers/nodes/types/Sop";
class ExporterUSDZSopParamsConfig extends BaseExporterSopParamsConfig {
}
const ParamsConfig = new ExporterUSDZSopParamsConfig();
export class ExporterUSDZSopNode extends ExporterSopNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return SopExporter.EXPORTER_USDZ;
  }
  fileExtension() {
    return "usdz";
  }
  createBlob() {
    return new Promise(async (resolve) => {
      const sceneData = await this._prepareScene();
      if (!sceneData) {
        return;
      }
      const { scene, objects } = sceneData;
      const exporter = new USDZExporter();
      const result = await exporter.parse(scene);
      console.log(result);
      this._handleResult(result, objects, resolve);
    });
  }
}
