// @polygonjs/polygonjs/dist/src/core/animation/Constant.js
"use strict";
export var AnimNodeEasing = /* @__PURE__ */ ((AnimNodeEasing2) => {
  AnimNodeEasing2["NONE"] = "none";
  AnimNodeEasing2["POWER1"] = "power1";
  AnimNodeEasing2["POWER2"] = "power2";
  AnimNodeEasing2["POWER3"] = "power3";
  AnimNodeEasing2["POWER4"] = "power4";
  AnimNodeEasing2["BACK"] = "back";
  AnimNodeEasing2["ELASTIC"] = "elastic";
  AnimNodeEasing2["BOUNCE"] = "bounce";
  AnimNodeEasing2["SLOW"] = "slow";
  AnimNodeEasing2["STEPS"] = "steps";
  AnimNodeEasing2["CIRC"] = "circ";
  AnimNodeEasing2["EXPO"] = "expo";
  AnimNodeEasing2["SINE"] = "sine";
  return AnimNodeEasing2;
})(AnimNodeEasing || {});
export const EASINGS = [
  "none" /* NONE */,
  "power1" /* POWER1 */,
  "power2" /* POWER2 */,
  "power3" /* POWER3 */,
  "power4" /* POWER4 */,
  "back" /* BACK */,
  "elastic" /* ELASTIC */,
  "bounce" /* BOUNCE */,
  "slow" /* SLOW */,
  "steps" /* STEPS */,
  "circ" /* CIRC */,
  "expo" /* EXPO */,
  "sine" /* SINE */
];
export var InOutMode = /* @__PURE__ */ ((InOutMode2) => {
  InOutMode2["IN"] = "in";
  InOutMode2["OUT"] = "out";
  InOutMode2["IN_OUT"] = "inOut";
  return InOutMode2;
})(InOutMode || {});
export const IN_OUT_MODES = ["in" /* IN */, "out" /* OUT */, "inOut" /* IN_OUT */];
