// @polygonjs/polygonjs/dist/src/engine/nodes/js/code/assemblers/instanceBuilder/InstanceBuilderAssemblerCommon.js
"use strict";
export var InstanceBuilderAssemblerConstant = /* @__PURE__ */ ((InstanceBuilderAssemblerConstant2) => {
  InstanceBuilderAssemblerConstant2["POINT_CONTAINER"] = "pointContainer";
  InstanceBuilderAssemblerConstant2["INSTANCE_POSITION"] = "pointContainer.instancePosition";
  InstanceBuilderAssemblerConstant2["INSTANCE_QUATERNION"] = "pointContainer.instanceQuaternion";
  InstanceBuilderAssemblerConstant2["INSTANCE_SCALE"] = "pointContainer.instanceScale";
  InstanceBuilderAssemblerConstant2["PTNUM"] = "pointContainer.ptnum";
  InstanceBuilderAssemblerConstant2["OBJNUM"] = "pointContainer.objnum";
  InstanceBuilderAssemblerConstant2["ATTRIBUTES_DICT"] = "attributesDict";
  InstanceBuilderAssemblerConstant2["OBJECT_3D"] = "null";
  InstanceBuilderAssemblerConstant2["MATERIAL"] = "null";
  return InstanceBuilderAssemblerConstant2;
})(InstanceBuilderAssemblerConstant || {});
export var InstanceVariable = /* @__PURE__ */ ((InstanceVariable2) => {
  InstanceVariable2["INSTANCE_POSITION"] = "instancePosition";
  InstanceVariable2["INSTANCE_QUATERNION"] = "instanceQuaternion";
  InstanceVariable2["INSTANCE_SCALE"] = "instanceScale";
  InstanceVariable2["PTNUM"] = "ptnum";
  InstanceVariable2["OBJNUM"] = "objnum";
  return InstanceVariable2;
})(InstanceVariable || {});
