// @polygonjs/polygonjs/dist/src/engine/nodes/gl/SDFHorseShoe.js
"use strict";
import { ThreeToGl } from "../../../core/ThreeToGl";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { GlConnectionPointType, GlConnectionPoint } from "../utils/io/connections/Gl";
import { BaseSDFGlNode } from "./_BaseSDF";
import { GlType } from "../../poly/registers/nodes/types/Gl";
const OUTPUT_NAME = "float";
class SDFHorseShoeGlParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    this.position = ParamConfig.VECTOR3([0, 0, 0], { hidden: true });
    this.center = ParamConfig.VECTOR3([0, 0, 0]);
    this.angle = ParamConfig.FLOAT(1, {
      range: [0, Math.PI],
      rangeLocked: [true, true],
      step: 1e-5
    });
    this.radius = ParamConfig.FLOAT(1);
    this.length = ParamConfig.FLOAT(1);
    this.thickness = ParamConfig.FLOAT(0.1);
    this.width = ParamConfig.FLOAT(0.1);
  }
}
const ParamsConfig = new SDFHorseShoeGlParamsConfig();
export class SDFHorseShoeGlNode extends BaseSDFGlNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return GlType.SDF_HORSE_SHOE;
  }
  initializeNode() {
    super.initializeNode();
    this.io.outputs.setNamedOutputConnectionPoints([
      new GlConnectionPoint(OUTPUT_NAME, GlConnectionPointType.FLOAT)
    ]);
  }
  setLines(shadersCollectionController) {
    const position = this.position();
    const center = ThreeToGl.vector3(this.variableForInputParam(this.p.center));
    const angle = ThreeToGl.float(this.variableForInputParam(this.p.angle));
    const radius = ThreeToGl.float(this.variableForInputParam(this.p.radius));
    const length = ThreeToGl.float(this.variableForInputParam(this.p.length));
    const thickness = ThreeToGl.float(this.variableForInputParam(this.p.thickness));
    const width = ThreeToGl.float(this.variableForInputParam(this.p.width));
    const float = this.glVarName(OUTPUT_NAME);
    const bodyLine = `float ${float} = sdHorseshoe(${position} - ${center}, ${angle}, ${radius}, ${length}, vec2(${thickness}, ${width}))`;
    shadersCollectionController.addBodyLines(this, [bodyLine]);
    this._addSDFMethods(shadersCollectionController);
  }
}
