// @polygonjs/polygonjs/dist/src/engine/nodes/gl/SDFTriangle.js
"use strict";
import { ThreeToGl } from "../../../../src/core/ThreeToGl";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { GlConnectionPointType, GlConnectionPoint } from "../utils/io/connections/Gl";
import { BaseSDFGlNode } from "./_BaseSDF";
import { GlType } from "../../poly/registers/nodes/types/Gl";
const OUTPUT_NAME = "float";
class SDFTriangleGlParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    this.position = ParamConfig.VECTOR3([0, 0, 0], { hidden: true });
    this.center = ParamConfig.VECTOR3([0, 0, 0]);
    this.a = ParamConfig.VECTOR3([0, 0, 0]);
    this.b = ParamConfig.VECTOR3([0, 1, 0]);
    this.c = ParamConfig.VECTOR3([0, 0, 1]);
    this.thickness = ParamConfig.FLOAT(0.1);
  }
}
const ParamsConfig = new SDFTriangleGlParamsConfig();
export class SDFTriangleGlNode extends BaseSDFGlNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return GlType.SDF_TRIANGLE;
  }
  initializeNode() {
    super.initializeNode();
    this.io.outputs.setNamedOutputConnectionPoints([
      new GlConnectionPoint(OUTPUT_NAME, GlConnectionPointType.FLOAT)
    ]);
  }
  setLines(shadersCollectionController) {
    const position = this.position();
    const center = ThreeToGl.vector3(this.variableForInputParam(this.p.center));
    const a = ThreeToGl.float(this.variableForInputParam(this.p.a));
    const b = ThreeToGl.float(this.variableForInputParam(this.p.b));
    const c = ThreeToGl.float(this.variableForInputParam(this.p.c));
    const thickness = ThreeToGl.float(this.variableForInputParam(this.p.thickness));
    const float = this.glVarName(OUTPUT_NAME);
    const bodyLine = `float ${float} = udTriangle(${position} - ${center}, ${a}, ${b}, ${c}, ${thickness})`;
    shadersCollectionController.addBodyLines(this, [bodyLine]);
    this._addSDFMethods(shadersCollectionController);
  }
}
