// @polygonjs/polygonjs/dist/src/engine/nodes/gl/SDFTube.js
"use strict";
import { ThreeToGl } from "../../../../src/core/ThreeToGl";
import { NodeParamsConfig, ParamConfig } from "../utils/params/ParamsConfig";
import { GlConnectionPointType, GlConnectionPoint } from "../utils/io/connections/Gl";
import { BaseSDFGlNode } from "./_BaseSDF";
import { isBooleanTrue } from "../../../core/Type";
import { GlType } from "../../poly/registers/nodes/types/Gl";
const OUTPUT_NAME = "float";
class SDFTubeGlParamsConfig extends NodeParamsConfig {
  constructor() {
    super(...arguments);
    this.position = ParamConfig.VECTOR3([0, 0, 0], { hidden: true });
    this.center = ParamConfig.VECTOR3([0, 0, 0]);
    this.capped = ParamConfig.BOOLEAN(1);
    this.radius = ParamConfig.FLOAT(0.1);
    this.height = ParamConfig.FLOAT(1, {
      visibleIf: { capped: 1 }
    });
  }
}
const ParamsConfig = new SDFTubeGlParamsConfig();
export class SDFTubeGlNode extends BaseSDFGlNode {
  constructor() {
    super(...arguments);
    this.paramsConfig = ParamsConfig;
  }
  static type() {
    return GlType.SDF_TUBE;
  }
  initializeNode() {
    super.initializeNode();
    this.io.connection_points.spare_params.setInputlessParamNames(["capped"]);
    this.io.outputs.setNamedOutputConnectionPoints([
      new GlConnectionPoint(OUTPUT_NAME, GlConnectionPointType.FLOAT)
    ]);
  }
  setLines(shadersCollectionController) {
    const position = this.position();
    const center = ThreeToGl.vector3(this.variableForInputParam(this.p.center));
    const radius = ThreeToGl.float(this.variableForInputParam(this.p.radius));
    const float = this.glVarName(OUTPUT_NAME);
    if (isBooleanTrue(this.pv.capped)) {
      const height = ThreeToGl.float(this.variableForInputParam(this.p.height));
      const bodyLine = `float ${float} = sdTubeCapped(${position} - ${center}, ${height}, ${radius})`;
      shadersCollectionController.addBodyLines(this, [bodyLine]);
    } else {
      const bodyLine = `float ${float} = sdTube(${position} - ${center}, ${radius})`;
      shadersCollectionController.addBodyLines(this, [bodyLine]);
    }
    this._addSDFMethods(shadersCollectionController);
  }
}
